{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "from Bio import SeqIO\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notebook to unpack the compressed genome and cds files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Unpack the nucleic acid fasta files into data/fna directoty.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home4/youn01f/Desktop/workspace/newData/fasta/virushostdb.genomic.fna.gz\n",
      "fna/AB007990.fasta is the 0 file\n",
      "fna/AB572858.fasta is the 100 file\n",
      "fna/AF109399.fasta is the 200 file\n",
      "fna/AJ132574.fasta is the 300 file\n",
      "fna/AM744992.fasta is the 400 file\n",
      "fna/AM774357.fasta is the 500 file\n",
      "fna/AY593796.fasta is the 600 file\n",
      "fna/HRVACG.fasta is the 700 file\n",
      "fna/DQ315364.fasta is the 800 file\n",
      "fna/EF067325.fasta is the 900 file\n",
      "fna/EU001278.fasta is the 1000 file\n",
      "fna/FJ025904.fasta is the 1100 file\n",
      "fna/FJ445168.fasta is the 1200 file\n",
      "fna/FM955839.fasta is the 1300 file\n",
      "fna/GU129016.fasta is the 1400 file\n",
      "fna/HM140370.fasta is the 1500 file\n",
      "fna/F1CCG.fasta is the 1600 file\n",
      "fna/JN400604.fasta is the 1700 file\n",
      "fna/JN819506.fasta is the 1800 file\n",
      "fna/JQ896627.fasta is the 1900 file\n",
      "fna/JX483874.fasta is the 2000 file\n",
      "fna/KC175383.fasta is the 2100 file\n",
      "fna/KC815311.fasta is the 2200 file\n",
      "fna/KF626665.fasta is the 2300 file\n",
      "fna/KJ407073.fasta is the 2400 file\n",
      "fna/KM225265.fasta is the 2500 file\n",
      "fna/KP687432.fasta is the 2600 file\n",
      "fna/NC_001265.fasta is the 2700 file\n",
      "fna/NC_001492.fasta is the 2800 file\n",
      "fna/NC_001676.fasta is the 2900 file\n",
      "fna/NC_001925.fasta is the 3000 file\n",
      "fna/NC_002206.fasta is the 3100 file\n",
      "fna/NC_002691.fasta is the 3200 file\n",
      "fna/NC_003345.fasta is the 3300 file\n",
      "fna/NC_003531.fasta is the 3400 file\n",
      "fna/NC_003664.fasta is the 3500 file\n",
      "fna/NC_003780.fasta is the 3600 file\n",
      "fna/NC_003899.fasta is the 3700 file\n",
      "fna/NC_004157.fasta is the 3800 file\n",
      "fna/NC_004440.fasta is the 3900 file\n",
      "fna/NC_004724.fasta is the 4000 file\n",
      "fna/NC_005136.fasta is the 4100 file\n",
      "fna/NC_005817.fasta is the 4200 file\n",
      "fna/NC_006013.fasta is the 4300 file\n",
      "fna/NC_006507.fasta is the 4400 file\n",
      "fna/NC_006998.fasta is the 4500 file\n",
      "fna/NC_007359.fasta is the 4600 file\n",
      "fna/NC_007591.fasta is the 4700 file\n",
      "fna/NC_007915.fasta is the 4800 file\n",
      "fna/NC_008250.fasta is the 4900 file\n",
      "fna/NC_008732.fasta is the 5000 file\n",
      "fna/NC_008924.fasta is the 5100 file\n",
      "fna/NC_009233.fasta is the 5200 file\n",
      "fna/NC_009758.fasta is the 5300 file\n",
      "fna/NC_010294.fasta is the 5400 file\n",
      "fna/NC_010707.fasta is the 5500 file\n",
      "fna/NC_011056.fasta is the 5600 file\n",
      "fna/NC_011544.fasta is the 5700 file\n",
      "fna/NC_012537.fasta is the 5800 file\n",
      "fna/NC_013060.fasta is the 5900 file\n",
      "fna/NC_013639.fasta is the 6000 file\n",
      "fna/NC_014322.fasta is the 6100 file\n",
      "fna/NC_014662.fasta is the 6200 file\n",
      "fna/NC_015210.fasta is the 6300 file\n",
      "fna/NC_015506.fasta is the 6400 file\n",
      "fna/NC_016404.fasta is the 6500 file\n",
      "fna/NC_016996.fasta is the 6600 file\n",
      "fna/NC_018173.fasta is the 6700 file\n",
      "fna/NC_018837.fasta is the 6800 file\n",
      "fna/NC_019498.fasta is the 6900 file\n",
      "fna/NC_019919.fasta is the 7000 file\n",
      "fna/NC_020486.fasta is the 7100 file\n",
      "fna/NC_021147.fasta is the 7200 file\n",
      "fna/NC_021546.fasta is the 7300 file\n",
      "fna/NC_021868.fasta is the 7400 file\n",
      "fna/NC_022563.fasta is the 7500 file\n",
      "fna/NC_022966.fasta is the 7600 file\n",
      "fna/NC_023443.fasta is the 7700 file\n",
      "fna/NC_023637.fasta is the 7800 file\n",
      "fna/NC_023848.fasta is the 7900 file\n",
      "fna/NC_024150.fasta is the 8000 file\n",
      "fna/NC_024491.fasta is the 8100 file\n",
      "fna/NC_024915.fasta is the 8200 file\n",
      "fna/NC_025416.fasta is the 8300 file\n",
      "fna/NC_025716.fasta is the 8400 file\n",
      "fna/NC_026015.fasta is the 8500 file\n",
      "fna/NC_026470.fasta is the 8600 file\n",
      "fna/NC_026797.fasta is the 8700 file\n",
      "fna/NC_027330.fasta is the 8800 file\n",
      "fna/NC_027552.fasta is the 8900 file\n",
      "fna/NC_027802.fasta is the 9000 file\n",
      "fna/NC_028136.fasta is the 9100 file\n",
      "fna/NC_028470.fasta is the 9200 file\n",
      "fna/NC_028758.fasta is the 9300 file\n",
      "fna/NC_028868.fasta is the 9400 file\n",
      "fna/NC_028973.fasta is the 9500 file\n",
      "fna/NC_029078.fasta is the 9600 file\n",
      "fna/NC_029580.fasta is the 9700 file\n",
      "fna/NC_029921.fasta is the 9800 file\n",
      "fna/NC_030290.fasta is the 9900 file\n",
      "fna/NC_030698.fasta is the 10000 file\n",
      "fna/NC_030934.fasta is the 10100 file\n",
      "fna/NC_031083.fasta is the 10200 file\n",
      "fna/NC_031256.fasta is the 10300 file\n",
      "fna/NC_031692.fasta is the 10400 file\n",
      "fna/NC_032118.fasta is the 10500 file\n",
      "fna/NC_032219.fasta is the 10600 file\n",
      "fna/NC_032440.fasta is the 10700 file\n",
      "fna/NC_032541.fasta is the 10800 file\n",
      "fna/NC_032641.fasta is the 10900 file\n",
      "fna/NC_032816.fasta is the 11000 file\n",
      "fna/NC_032916.fasta is the 11100 file\n",
      "fna/NC_033016.fasta is the 11200 file\n",
      "fna/NC_033116.fasta is the 11300 file\n",
      "fna/NC_033216.fasta is the 11400 file\n",
      "fna/NC_033316.fasta is the 11500 file\n",
      "fna/NC_033496.fasta is the 11600 file\n",
      "fna/NC_033786.fasta is the 11700 file\n",
      "fna/NC_034249.fasta is the 11800 file\n",
      "fna/NC_034486.fasta is the 11900 file\n",
      "fna/NC_034631.fasta is the 12000 file\n",
      "fna/NC_035451.fasta is the 12100 file\n",
      "fna/NC_036615.fasta is the 12200 file\n",
      "fna/NC_038275.fasta is the 12300 file\n",
      "fna/NC_038375.fasta is the 12400 file\n",
      "fna/NC_038475.fasta is the 12500 file\n",
      "fna/NC_038575.fasta is the 12600 file\n",
      "fna/NC_038675.fasta is the 12700 file\n",
      "fna/NC_038775.fasta is the 12800 file\n",
      "fna/NC_038875.fasta is the 12900 file\n",
      "fna/NC_038975.fasta is the 13000 file\n",
      "fna/NC_039075.fasta is the 13100 file\n",
      "fna/BLU20621.fasta is the 13200 file\n",
      "13290 files extracted\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "fastapath = \"fna/\"\n",
    "inputfile = \"fna/virushostdb.genomic.fna.gz\"\n",
    "print (inputfile)\n",
    "with gzip.open(inputfile, \"rt\") as handle:\n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "        fileout = os.path.join(fastapath, record.id +\".fasta\")\n",
    "        with open(fileout, \"w\") as output_handle:\n",
    "            SeqIO.write(record, output_handle, \"fasta\")\n",
    "            if count%500 == 0:\n",
    "                print (f'file {count}: {fileout} ')           \n",
    "        count += 1\n",
    "print (f'{count} files extracted')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Unpack the faa files into data/faa\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home4/youn01f/Desktop/workspace/newData/faa/virushostdb.cds.faa.gz\n",
      "output genome AB007990 6 AB012574\n",
      "output genome AB012574 7 AB024414\n",
      "output genome AB024414 65 AB027020\n",
      "output genome AB027020 8 AB027021\n",
      "output genome AB027021 8 AB032553\n",
      "output genome AB032553 3 AB042808\n",
      "output genome AB042808 3 AB050936\n",
      "output genome AB050936 8 AB073912\n",
      "output genome AB073912 3 AB090161\n",
      "output genome AB090161 2 AB110218\n",
      "output genome AB110218 6 AB162141\n",
      "output genome AB162141 6 AB176642\n",
      "output genome AB176642 3 AB187514\n",
      "output genome AB187514 3 AB187566\n",
      "output genome AB187566 2 AB192965\n",
      "output genome AB192965 6 AB192966\n",
      "output genome AB192966 6 AB194796\n",
      "output genome AB194796 1 AB205396\n",
      "output genome AB205396 1 AB211993\n",
      "output genome AB211993 7 AB220921\n",
      "output genome AB220921 3 AB236321\n",
      "output genome AB236321 6 AB236323\n",
      "output genome AB236323 6 AB236325\n",
      "output genome AB236325 6 AB236326\n",
      "output genome AB236326 1 AB236327\n",
      "output genome AB236327 1 AB252582\n",
      "output genome AB252582 1 AB300463\n",
      "output genome AB300463 6 AB306314\n",
      "output genome AB306314 6 AB306522\n",
      "output genome AB306522 1 AB331650\n",
      "output genome AB331650 6 AB331651\n",
      "output genome AB331651 7 AB361563\n",
      "output genome AB361563 7 AB365435\n",
      "output genome AB365435 3 AB374228\n",
      " number of CDS 250 number of genomes 35\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "outpath = \"faa/\"\n",
    "inputfile = \"faa/virushostdb.cds.faa.gz\"\n",
    "print (inputfile)\n",
    "path = 'fasta/'\n",
    "seq_list = []\n",
    "old_refid = 'xxxx'\n",
    "n_records = 0\n",
    "all_records = []\n",
    "all_n_records = []\n",
    "\n",
    "with gzip.open(inputfile, \"rt\") as handle:\n",
    "    \n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "        \n",
    "        count += 1\n",
    "        #Parse out the sequence id             \n",
    "        # example        BAA82924.1|/gene=\"ORF 36\",/product=\"UL28 product homolog\"||AB024414|[49188:51624](-)\n",
    "        refid = record.description.split('|')[-2]\n",
    "\n",
    "        if count == 1:  # for the first record  in fileonly\n",
    "            seq_list.append(refid)\n",
    "            old_refid = refid\n",
    "\n",
    "        if refid == old_refid: # another record for same refseq\n",
    "            all_records.append(record)\n",
    "            n_records += 1\n",
    "\n",
    "        else:\n",
    "        # CDS from a new species , write out all_records to a file\n",
    "\n",
    "            all_n_records.append(n_records)\n",
    "            SeqIO.write(all_records, f\"{outpath}{old_refid}.fasta\", \"fasta\")  \n",
    "            seq_list.append(refid)\n",
    "            old_refid = refid\n",
    "            n_records = 1\n",
    "            all_records = [record]\n",
    "            if count%500 == 0:\n",
    "                print (f'Working on {old_refid} and the {count} CDS record')   \n",
    "    \n",
    "print (' number of CDS', count, 'number of genomes',len(seq_list))\n",
    "\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
