{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "import os\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import random\n",
    "import numpy as np\n",
    "import timeit\n",
    "from collections import defaultdict, Counter\n",
    "from pathlib import Path\n",
    "from Bio import SeqIO\n",
    "import sys\n",
    "lib_dir = '../mylibs'\n",
    "if lib_dir not in sys.path:\n",
    "    sys.path.append(lib_dir)\n",
    "    \n",
    "import vhdb as vhdb\n",
    "import features_predict as fp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "filein = '../inputs/VHDB_25_1_2019.p'\n",
    "with open(filein, 'rb') as f:\n",
    "    V_H = pickle.load( f)\n",
    "hosts = V_H.hosts\n",
    "viruses = V_H.viruses\n",
    "subspecies =V_H.subspecies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results in Euk_RNA_results.csv\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Input file\n",
    "#subsets = [(label,taxlevel,pool,taxlevelpool),...]\n",
    "subsetfile = '../inputs/Euk_RNA.csv'\n",
    "baltimore = 'all'\n",
    "vhdbfile = '../inputs/VHDB_25_1_2019.p'\n",
    "\n",
    "\n",
    "# Everthing in the same order as the features list\n",
    "\n",
    "features = ['DNA','AA','PC','Domains']\n",
    "# filepath for the fasta, faa and domain files\n",
    "filepaths = [\n",
    "                '/home4/youn01f/Desktop/workspace/newData/fasta',\n",
    "                '/home4/youn01f/Desktop/workspace/newData/faa','/home4/youn01f/Desktop/workspace/newData/faa',\n",
    "                '/home4/youn01f/Desktop/workspace/newData/pfs']\n",
    "file_exts = ['fasta','fasta','fasta','pfs']#,'fasta',fasta]\n",
    "\n",
    "# A list of kmers for each feature set to be tested\n",
    "kmer_lists = [[1, 2,3,4,5,6,7,8,9], # dna \n",
    "              [1, 2, 3,4], # aa\n",
    "              [1, 2,3,4,5,6] ,    #pc\n",
    "             [0]] #domains\n",
    "\n",
    "# symbol dictionaries \n",
    "na_dict = {'mod':4,'a':0,'c':1,'g':2,'t':3}\n",
    "aa_dict = {'mod':20 ,'A':0,'C':1,'D':2,'E':3,'F':4,'G':5,'H':6,'I':7,'K':8,'L':9,'M':10,'N':11,\n",
    "              'P':12,'Q':13,'R':14,'S':15,'T':16,'V':17,'W':18,'Y':19}\n",
    "pc_dict = {'mod':7, 'C':0,\n",
    "            'A':1,'G':1,'V':1,\n",
    "           'I':2,'L':2,'F':2,'P':2,\n",
    "           'M':3,'S':3,'T':3,'Y':3,\n",
    "           'H':4,'N':4,'Q':4,'W':4,\n",
    "           'R':5,'K':5,\n",
    "           'D':6,'E':6}\n",
    "symbol_dicts = [ na_dict,aa_dict,pc_dict, {}]\n",
    "\n",
    "classifiers = ['SVM_lin']\n",
    "\n",
    "# Output file for the results\n",
    "\n",
    "results_file = f'{Path(subsetfile).stem}_results.csv'\n",
    "print( f'Results in {results_file}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "51"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "label_info = pd.read_csv(subsetfile)\n",
    "subsets = label_info.apply(tuple, axis=1).tolist()\n",
    "len(subsets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Bos', 'genus', 'Bovidae', 'family', 'RNA')\n",
      " size of training 172 and test set   40\n",
      "('Ovis aries', 'species', 'Bovidae', 'family', 'RNA')\n",
      " size of training 42 and test set   14\n",
      "('Bovidae', 'family', 'Mammalia', 'class', 'RNA')\n",
      " size of training 224 and test set   58\n",
      "('Bos taurus', 'species', 'Bovidae', 'family', 'RNA')\n",
      " size of training 161 and test set   39\n",
      "('Sus scrofa', 'species', 'Mammalia', 'class', 'RNA')\n",
      " size of training 169 and test set   39\n",
      "('Suidae', 'family', 'Mammalia', 'class', 'RNA')\n",
      " size of training 173 and test set   41\n",
      "('Vespertilionidae', 'family', 'Chiroptera', 'order', 'RNA')\n",
      " size of training 56 and test set   18\n",
      "('Pteropodidae', 'family', 'Chiroptera', 'order', 'RNA')\n",
      " size of training 68 and test set   20\n",
      "('Chiroptera', 'order', 'Mammalia', 'class', 'RNA')\n",
      " size of training 250 and test set   64\n",
      "('Equus', 'genus', 'Mammalia', 'class', 'RNA')\n",
      " size of training 44 and test set   16\n",
      "('Equidae', 'family', 'Mammalia', 'class', 'RNA')\n",
      " size of training 46 and test set   16\n",
      "('Perissodactyla', 'order', 'Mammalia', 'class', 'RNA')\n",
      " size of training 46 and test set   16\n",
      "('Felis catus', 'species', 'Carnivora', 'order', 'RNA')\n",
      " size of training 58 and test set   18\n",
      "('Felis', 'genus', 'Carnivora', 'order', 'RNA')\n",
      " size of training 58 and test set   18\n",
      "('Felidae', 'family', 'Carnivora', 'order', 'RNA')\n",
      " size of training 68 and test set   20\n",
      "('Carnivora', 'order', 'Mammalia', 'class', 'RNA')\n",
      " size of training 150 and test set   34\n",
      "('Homo sapiens', 'species', 'Primates', 'order', 'RNA')\n",
      " size of training 402 and test set   98\n",
      "('Homo', 'genus', 'Primates', 'order', 'RNA')\n",
      " size of training 402 and test set   98\n",
      "('Hominidae', 'family', 'Primates', 'order', 'RNA')\n",
      " size of training 402 and test set   98\n",
      "('Macaca', 'genus', 'Cercopithecidae', 'family', 'RNA')\n",
      " size of training 48 and test set   16\n",
      "('Chlorocebus aethiops', 'species', 'Cercopithecidae', 'family', 'RNA')\n",
      " size of training 68 and test set   20\n",
      "('Chlorocebus', 'genus', 'Cercopithecidae', 'family', 'RNA')\n",
      " size of training 76 and test set   20\n",
      "('Cercopithecidae', 'family', 'Primates', 'order', 'RNA')\n",
      " size of training 157 and test set   35\n",
      "('Primates', 'order', 'Mammalia', 'class', 'RNA')\n",
      " size of training 402 and test set   98\n",
      "('Mus musculus', 'species', 'Muridae', 'family', 'RNA')\n",
      " size of training 178 and test set   44\n",
      "('Mus', 'genus', 'Muridae', 'family', 'RNA')\n",
      " size of training 192 and test set   50\n",
      "('Rattus', 'genus', 'Muridae', 'family', 'RNA')\n",
      " size of training 48 and test set   16\n",
      "('Muridae', 'family', 'Rodentia', 'order', 'RNA')\n",
      " size of training 275 and test set   71\n",
      "('Cricetidae', 'family', 'Rodentia', 'order', 'RNA')\n",
      " size of training 134 and test set   34\n",
      "('Rodentia', 'order', 'Mammalia', 'class', 'RNA')\n",
      " size of training 402 and test set   98\n",
      "('Mammalia', 'class', 'Chordata', 'phylum', 'RNA')\n",
      " size of training 402 and test set   98\n",
      "('Gallus gallus', 'species', 'Aves', 'class', 'RNA')\n",
      " size of training 68 and test set   20\n",
      "('Phasianidae', 'family', 'Aves', 'class', 'RNA')\n",
      " size of training 96 and test set   20\n",
      "('Galliformes', 'order', 'Aves', 'class', 'RNA')\n",
      " size of training 107 and test set   25\n",
      "('Anas platyrhynchos', 'species', 'Aves', 'class', 'RNA')\n",
      " size of training 42 and test set   14\n",
      "('Anas', 'genus', 'Aves', 'class', 'RNA')\n",
      " size of training 46 and test set   16\n",
      "('Anatidae', 'family', 'Aves', 'class', 'RNA')\n",
      " size of training 74 and test set   20\n",
      "('Aves', 'class', 'Chordata', 'phylum', 'RNA')\n",
      " size of training 297 and test set   73\n",
      "('Chordata', 'phylum', 'Metazoa', 'kingdom', 'RNA')\n",
      " size of training 402 and test set   98\n",
      "('Hymenoptera', 'order', 'Insecta', 'class', 'RNA')\n",
      " size of training 58 and test set   18\n",
      "('Aedes', 'genus', 'Culicidae', 'family', 'RNA')\n",
      " size of training 74 and test set   20\n",
      "('Culex', 'genus', 'Culicidae', 'family', 'RNA')\n",
      " size of training 118 and test set   30\n",
      "('Culicidae', 'family', 'Diptera', 'order', 'RNA')\n",
      " size of training 277 and test set   71\n",
      "('Diptera', 'order', 'Insecta', 'class', 'RNA')\n",
      " size of training 397 and test set   95\n",
      "('Lepidoptera', 'order', 'Insecta', 'class', 'RNA')\n",
      " size of training 62 and test set   18\n",
      "('Hemiptera', 'order', 'Insecta', 'class', 'RNA')\n",
      " size of training 82 and test set   18\n",
      "('Insecta', 'class', 'Arthropoda', 'phylum', 'RNA')\n",
      " size of training 402 and test set   98\n",
      "('Ixodidae', 'family', 'Arthropoda', 'phylum', 'RNA')\n",
      " size of training 76 and test set   20\n",
      "('Ixodida', 'order', 'Arthropoda', 'phylum', 'RNA')\n",
      " size of training 101 and test set   21\n",
      "('Arachnida', 'class', 'Arthropoda', 'phylum', 'RNA')\n",
      " size of training 111 and test set   29\n",
      "('Arthropoda', 'phylum', 'Metazoa', 'kingdom', 'RNA')\n",
      " size of training 402 and test set   98\n"
     ]
    }
   ],
   "source": [
    "for subset in subsets:\n",
    "    print (subset)\n",
    "    class_data,n = fp.get_class_lists(subset,viruses,hosts)\n",
    "    datasets = fp.split_data(viruses,class_data,n)\n",
    "    (label,label_tax,pool,pool_tax,baltimore) = subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "min(500,250,251)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_kmers (sequences,k,symbol_dict):\n",
    "    if k == 0: # domainsdom_list = list({dom for s in sequence.values() for dom in s})\n",
    "        \n",
    "        X,seq_index,f_index = get_domains(sequences,symbol_dict)\n",
    "    else:  # everything else\n",
    "        vocab_len =  symbol_dict['mod'] **k\n",
    "        f_index = get_feature_names(symbol_dict,k)\n",
    "        seq_index = []\n",
    "        X = np.zeros((len(sequences) ,vocab_len))\n",
    "\n",
    "        for i,(accn, seq) in enumerate (sequences.items()):\n",
    "            X[i] = get_word_frequ (seq,k, vocab_len,symbol_dict)\n",
    "            seq_index.append(accn)\n",
    "# normalise for length of genome(s)\n",
    "    XN = np.divide(X,X.sum(axis = 1)[:,None])\n",
    "    print (np.shape(XN)) \n",
    "    return XN,seq_index,f_index\n",
    "\n",
    "def get_domains(sequence,dom_list):\n",
    "    \"\"\"input: sequences dictionary of training and test sequence dicts\"\"\"\n",
    "    seq_index = []\n",
    "    X = np.zeros((len(sequence) ,len(dom_list)))\n",
    "    for i,(accn, seq) in enumerate (sequence.items()):\n",
    "        for dom in seq:\n",
    "            j = dom_list.index(dom)\n",
    "            X[i,j] += 1\n",
    "        seq_index.append(accn)\n",
    "    return X,seq_index,dom_list\n",
    "\n",
    "def get_word_frequ(sequence,k,vocab_len,symbol_dict):\n",
    "    \"\"\"returns the word frequency array for all genomes for a virus \"\"\"\n",
    "    word_freqs = np.zeros(shape =(vocab_len),dtype=np.int32)\n",
    "    for i in range(len(sequence)-k+1):\n",
    "        word = sequence[i:i+k]\n",
    "        if check_word(word,symbol_dict):\n",
    "            kmer_index = patternToNumber (word,k, symbol_dict)\n",
    "            word_freqs [ kmer_index] += 1\n",
    "            # If this for DNA/RNA then add the k-mers complinent\n",
    "#             if symbol_dict['mod'] == 4:\n",
    "#                 word_freqs [patternToNumber(compliment(word),k,symbol_dict)] +=1   \n",
    "    return word_freqs\n",
    "\n",
    "def get_sequences(faapath,datasets,ext):\n",
    "    missing_files = []\n",
    "    v_with_missing_files = []\n",
    "    ext = f'.{ext}'\n",
    "    sequences = {'training':{},'test':{}}\n",
    "    for key,dataset in datasets.items():\n",
    "        for v,vdict in dataset.items():\n",
    "            sequence = defaultdict(str)\n",
    "            for ref in vdict['refseqs']:\n",
    "                filename = os.path.join(faapath, ref.strip()+ ext)\n",
    "                try:\n",
    "                    with open(filename, 'rt') as handle:\n",
    "                        if ext == '.fasta':\n",
    "                            for record in SeqIO.parse(handle, \"fasta\"):                   \n",
    "                                sequence[v] += str(record.seq)\n",
    "                        else: # pfs file - list of domains  \n",
    "                            line = handle.read()\n",
    "                            sequence[v] = [d for d in line.split()]\n",
    "                                \n",
    "                except IOError:\n",
    "                    missing_files.append((v,ref))\n",
    "                    \n",
    "            if len(sequence) > 0:\n",
    "                sequences[key].update(sequence)\n",
    "            else:\n",
    "                v_with_missing_files.append(v)\n",
    "    print ( f'viruses no files {len(set(v_with_missing_files))}')\n",
    "    return (sequences)\n",
    "\n",
    "\n",
    "def symbolToNumber (symbol,symbols):\n",
    "    return symbols[symbol]\n",
    "\n",
    "def patternToNumber (pattern,k,symbol_dict):\n",
    "    mod = symbol_dict['mod']\n",
    "    number = 0\n",
    "    for i in range (0,k):\n",
    "        n = symbolToNumber ( pattern [k-1-i],symbol_dict)\n",
    "        number += n* mod**i \n",
    "    return number\n",
    "\n",
    "def get_feature_names(symbols,k):\n",
    "    all_kmers = []\n",
    "    n2sym = { v:k for k,v in symbols.items()}\n",
    "    mod = symbols['mod']\n",
    "    for number in range(mod**k):\n",
    "        pattern = ''\n",
    "        for i in range (0,k):\n",
    "            n = number% mod\n",
    "            number = (number - n)/mod \n",
    "            symbol = n2sym[n]\n",
    "            pattern = symbol + pattern\n",
    "        all_kmers.append (pattern)\n",
    "    return all_kmers\n",
    "\n",
    "def check_word(mystring,symbols):\n",
    "    \"\"\"Check for illegal characters \"\"\"\n",
    "    return all(c in symbols for c in mystring)\n",
    "\n",
    "def compliment(dna):\n",
    "    \"\"\"Return the reverse compliment of a sequence \"\"\"\n",
    "    bp = { 'a':'t', 't':'a', 'c':'g','g':'c'}\n",
    "    revdna =''    \n",
    "    for i in range(len(dna)-1,-1,-1):\n",
    "         revdna += bp[dna[i]]\n",
    "    return (revdna)\n",
    "\n",
    "def  get_labels(label,v_index,dataset):      \n",
    "    for i,v in enumerate(v_index):\n",
    "        y = [1 if dataset[v]['label'] ==  label else 0 for v in v_index ] \n",
    "    return np.asarray (y)\n",
    "\n",
    "def get_feature_matrices(sequences,datasets,k,symbol_dict):\n",
    "   \n",
    "    if k == 0: # domainsdom_list = list({dom for s in sequence.values() for dom in s})\n",
    "        symbol_dict = list({dom for sequence in sequences.values() for s in sequence.values() for dom in s}) \n",
    "    print (f'X_training:Extacting  {feature}  of length  {k}  from   {len(sequences[\"training\"])} sequences')\n",
    "    X_train, seq_index, f_index = extract_kmers (sequences['training'],k,symbol_dict)\n",
    "    Y_train =  get_labels (label,seq_index,datasets['training'])\n",
    "    print (f'X_test: Extacting  {feature}  of length  {k}  from   {len(sequences[\"test\"])} sequences')\n",
    "    X_test, seq_index, f_index = extract_kmers (sequences['test'],k,symbol_dict)\n",
    "    Y_test =  get_labels (label,seq_index,datasets['test'])\n",
    "           \n",
    "    return X_train,X_test,Y_train,Y_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main Loop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " size of training 172 and test set   40\n",
      "Bos genus Bovidae family RNA\n",
      "0 [1, 2, 3, 4, 5, 6, 7, 8, 9] /home4/youn01f/Desktop/workspace/newData/fasta {'mod': 4, 'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
      "getting DNA sequences from /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "viruses no files 3\n",
      "X_training:Extacting  DNA  of length  1  from   170 sequences\n",
      "(170, 4)\n",
      "X_test: Extacting  DNA  of length  1  from   39 sequences\n",
      "(39, 4)\n",
      "(170, 4) (39, 4) (170,) (39,)\n",
      "{'AUC': 0.534, 'N in class': 106, 'Features': 'DNA', 'k': 1}\n",
      "X_training:Extacting  DNA  of length  2  from   170 sequences\n",
      "(170, 16)\n",
      "X_test: Extacting  DNA  of length  2  from   39 sequences\n",
      "(39, 16)\n",
      "(170, 16) (39, 16) (170,) (39,)\n",
      "{'AUC': 0.675, 'N in class': 106, 'Features': 'DNA', 'k': 2}\n",
      "X_training:Extacting  DNA  of length  3  from   170 sequences\n",
      "(170, 64)\n",
      "X_test: Extacting  DNA  of length  3  from   39 sequences\n",
      "(39, 64)\n",
      "(170, 64) (39, 64) (170,) (39,)\n",
      "{'AUC': 0.817, 'N in class': 106, 'Features': 'DNA', 'k': 3}\n",
      "X_training:Extacting  DNA  of length  4  from   170 sequences\n",
      "(170, 256)\n",
      "X_test: Extacting  DNA  of length  4  from   39 sequences\n",
      "(39, 256)\n",
      "(170, 256) (39, 256) (170,) (39,)\n",
      "{'AUC': 0.876, 'N in class': 106, 'Features': 'DNA', 'k': 4}\n",
      "X_training:Extacting  DNA  of length  5  from   170 sequences\n",
      "(170, 1024)\n",
      "X_test: Extacting  DNA  of length  5  from   39 sequences\n",
      "(39, 1024)\n",
      "(170, 1024) (39, 1024) (170,) (39,)\n",
      "{'AUC': 0.921, 'N in class': 106, 'Features': 'DNA', 'k': 5}\n",
      "X_training:Extacting  DNA  of length  6  from   170 sequences\n",
      "(170, 4096)\n",
      "X_test: Extacting  DNA  of length  6  from   39 sequences\n",
      "(39, 4096)\n",
      "(170, 4096) (39, 4096) (170,) (39,)\n",
      "{'AUC': 0.926, 'N in class': 106, 'Features': 'DNA', 'k': 6}\n",
      "X_training:Extacting  DNA  of length  7  from   170 sequences\n",
      "(170, 16384)\n",
      "X_test: Extacting  DNA  of length  7  from   39 sequences\n",
      "(39, 16384)\n",
      "(170, 16384) (39, 16384) (170,) (39,)\n",
      "{'AUC': 0.955, 'N in class': 106, 'Features': 'DNA', 'k': 7}\n",
      "X_training:Extacting  DNA  of length  8  from   170 sequences\n",
      "(170, 65536)\n",
      "X_test: Extacting  DNA  of length  8  from   39 sequences\n",
      "(39, 65536)\n",
      "(170, 65536) (39, 65536) (170,) (39,)\n",
      "{'AUC': 0.963, 'N in class': 106, 'Features': 'DNA', 'k': 8}\n",
      "X_training:Extacting  DNA  of length  9  from   170 sequences\n",
      "(170, 262144)\n",
      "X_test: Extacting  DNA  of length  9  from   39 sequences\n",
      "(39, 262144)\n",
      "(170, 262144) (39, 262144) (170,) (39,)\n",
      "{'AUC': 0.937, 'N in class': 106, 'Features': 'DNA', 'k': 9}\n",
      "1 [1, 2, 3, 4] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 20, 'A': 0, 'C': 1, 'D': 2, 'E': 3, 'F': 4, 'G': 5, 'H': 6, 'I': 7, 'K': 8, 'L': 9, 'M': 10, 'N': 11, 'P': 12, 'Q': 13, 'R': 14, 'S': 15, 'T': 16, 'V': 17, 'W': 18, 'Y': 19}\n",
      "getting AA sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "viruses no files 2\n",
      "X_training:Extacting  AA  of length  1  from   170 sequences\n",
      "(170, 20)\n",
      "X_test: Extacting  AA  of length  1  from   40 sequences\n",
      "(40, 20)\n",
      "(170, 20) (40, 20) (170,) (40,)\n",
      "{'AUC': 0.816, 'N in class': 106, 'Features': 'AA', 'k': 1}\n",
      "X_training:Extacting  AA  of length  2  from   170 sequences\n",
      "(170, 400)\n",
      "X_test: Extacting  AA  of length  2  from   40 sequences\n",
      "(40, 400)\n",
      "(170, 400) (40, 400) (170,) (40,)\n",
      "{'AUC': 0.864, 'N in class': 106, 'Features': 'AA', 'k': 2}\n",
      "X_training:Extacting  AA  of length  3  from   170 sequences\n",
      "(170, 8000)\n",
      "X_test: Extacting  AA  of length  3  from   40 sequences\n",
      "(40, 8000)\n",
      "(170, 8000) (40, 8000) (170,) (40,)\n",
      "{'AUC': 0.962, 'N in class': 106, 'Features': 'AA', 'k': 3}\n",
      "X_training:Extacting  AA  of length  4  from   170 sequences\n",
      "(170, 160000)\n",
      "X_test: Extacting  AA  of length  4  from   40 sequences\n",
      "(40, 160000)\n",
      "(170, 160000) (40, 160000) (170,) (40,)\n",
      "{'AUC': 0.947, 'N in class': 106, 'Features': 'AA', 'k': 4}\n",
      "2 [1, 2, 3, 4, 5, 6] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 7, 'C': 0, 'A': 1, 'G': 1, 'V': 1, 'I': 2, 'L': 2, 'F': 2, 'P': 2, 'M': 3, 'S': 3, 'T': 3, 'Y': 3, 'H': 4, 'N': 4, 'Q': 4, 'W': 4, 'R': 5, 'K': 5, 'D': 6, 'E': 6}\n",
      "getting PC sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "viruses no files 2\n",
      "X_training:Extacting  PC  of length  1  from   170 sequences\n",
      "(170, 7)\n",
      "X_test: Extacting  PC  of length  1  from   40 sequences\n",
      "(40, 7)\n",
      "(170, 7) (40, 7) (170,) (40,)\n",
      "{'AUC': 0.573, 'N in class': 106, 'Features': 'PC', 'k': 1}\n",
      "X_training:Extacting  PC  of length  2  from   170 sequences\n",
      "(170, 49)\n",
      "X_test: Extacting  PC  of length  2  from   40 sequences\n",
      "(40, 49)\n",
      "(170, 49) (40, 49) (170,) (40,)\n",
      "{'AUC': 0.836, 'N in class': 106, 'Features': 'PC', 'k': 2}\n",
      "X_training:Extacting  PC  of length  3  from   170 sequences\n",
      "(170, 343)\n",
      "X_test: Extacting  PC  of length  3  from   40 sequences\n",
      "(40, 343)\n",
      "(170, 343) (40, 343) (170,) (40,)\n",
      "{'AUC': 0.932, 'N in class': 106, 'Features': 'PC', 'k': 3}\n",
      "X_training:Extacting  PC  of length  4  from   170 sequences\n",
      "(170, 2401)\n",
      "X_test: Extacting  PC  of length  4  from   40 sequences\n",
      "(40, 2401)\n",
      "(170, 2401) (40, 2401) (170,) (40,)\n",
      "{'AUC': 0.914, 'N in class': 106, 'Features': 'PC', 'k': 4}\n",
      "X_training:Extacting  PC  of length  5  from   170 sequences\n",
      "(170, 16807)\n",
      "X_test: Extacting  PC  of length  5  from   40 sequences\n",
      "(40, 16807)\n",
      "(170, 16807) (40, 16807) (170,) (40,)\n",
      "{'AUC': 0.922, 'N in class': 106, 'Features': 'PC', 'k': 5}\n",
      "X_training:Extacting  PC  of length  6  from   170 sequences\n",
      "(170, 117649)\n",
      "X_test: Extacting  PC  of length  6  from   40 sequences\n",
      "(40, 117649)\n",
      "(170, 117649) (40, 117649) (170,) (40,)\n",
      "{'AUC': 0.924, 'N in class': 106, 'Features': 'PC', 'k': 6}\n",
      "3 [0] /home4/youn01f/Desktop/workspace/newData/pfs {}\n",
      "getting Domains sequences from /home4/youn01f/Desktop/workspace/newData/pfs\n",
      "viruses no files 3\n",
      "X_training:Extacting  Domains  of length  0  from   169 sequences\n",
      "(169, 182)\n",
      "X_test: Extacting  Domains  of length  0  from   40 sequences\n",
      "(40, 182)\n",
      "(169, 182) (40, 182) (169,) (40,)\n",
      "{'AUC': 0.854, 'N in class': 106, 'Features': 'Domains', 'k': 0}\n",
      " size of training 42 and test set   14\n",
      "Ovis aries species Bovidae family RNA\n",
      "0 [1, 2, 3, 4, 5, 6, 7, 8, 9] /home4/youn01f/Desktop/workspace/newData/fasta {'mod': 4, 'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
      "getting DNA sequences from /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "viruses no files 1\n",
      "X_training:Extacting  DNA  of length  1  from   42 sequences\n",
      "(42, 4)\n",
      "X_test: Extacting  DNA  of length  1  from   13 sequences\n",
      "(13, 4)\n",
      "(42, 4) (13, 4) (42,) (13,)\n",
      "{'AUC': 0.571, 'N in class': 28, 'Features': 'DNA', 'k': 1}\n",
      "X_training:Extacting  DNA  of length  2  from   42 sequences\n",
      "(42, 16)\n",
      "X_test: Extacting  DNA  of length  2  from   13 sequences\n",
      "(13, 16)\n",
      "(42, 16) (13, 16) (42,) (13,)\n",
      "{'AUC': 0.524, 'N in class': 28, 'Features': 'DNA', 'k': 2}\n",
      "X_training:Extacting  DNA  of length  3  from   42 sequences\n",
      "(42, 64)\n",
      "X_test: Extacting  DNA  of length  3  from   13 sequences\n",
      "(13, 64)\n",
      "(42, 64) (13, 64) (42,) (13,)\n",
      "{'AUC': 0.643, 'N in class': 28, 'Features': 'DNA', 'k': 3}\n",
      "X_training:Extacting  DNA  of length  4  from   42 sequences\n",
      "(42, 256)\n",
      "X_test: Extacting  DNA  of length  4  from   13 sequences\n",
      "(13, 256)\n",
      "(42, 256) (13, 256) (42,) (13,)\n",
      "{'AUC': 0.595, 'N in class': 28, 'Features': 'DNA', 'k': 4}\n",
      "X_training:Extacting  DNA  of length  5  from   42 sequences\n",
      "(42, 1024)\n",
      "X_test: Extacting  DNA  of length  5  from   13 sequences\n",
      "(13, 1024)\n",
      "(42, 1024) (13, 1024) (42,) (13,)\n",
      "{'AUC': 0.595, 'N in class': 28, 'Features': 'DNA', 'k': 5}\n",
      "X_training:Extacting  DNA  of length  6  from   42 sequences\n",
      "(42, 4096)\n",
      "X_test: Extacting  DNA  of length  6  from   13 sequences\n",
      "(13, 4096)\n",
      "(42, 4096) (13, 4096) (42,) (13,)\n",
      "{'AUC': 0.619, 'N in class': 28, 'Features': 'DNA', 'k': 6}\n",
      "X_training:Extacting  DNA  of length  7  from   42 sequences\n",
      "(42, 16384)\n",
      "X_test: Extacting  DNA  of length  7  from   13 sequences\n",
      "(13, 16384)\n",
      "(42, 16384) (13, 16384) (42,) (13,)\n",
      "{'AUC': 0.667, 'N in class': 28, 'Features': 'DNA', 'k': 7}\n",
      "X_training:Extacting  DNA  of length  8  from   42 sequences\n",
      "(42, 65536)\n",
      "X_test: Extacting  DNA  of length  8  from   13 sequences\n",
      "(13, 65536)\n",
      "(42, 65536) (13, 65536) (42,) (13,)\n",
      "{'AUC': 0.69, 'N in class': 28, 'Features': 'DNA', 'k': 8}\n",
      "X_training:Extacting  DNA  of length  9  from   42 sequences\n",
      "(42, 262144)\n",
      "X_test: Extacting  DNA  of length  9  from   13 sequences\n",
      "(13, 262144)\n",
      "(42, 262144) (13, 262144) (42,) (13,)\n",
      "{'AUC': 0.786, 'N in class': 28, 'Features': 'DNA', 'k': 9}\n",
      "1 [1, 2, 3, 4] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 20, 'A': 0, 'C': 1, 'D': 2, 'E': 3, 'F': 4, 'G': 5, 'H': 6, 'I': 7, 'K': 8, 'L': 9, 'M': 10, 'N': 11, 'P': 12, 'Q': 13, 'R': 14, 'S': 15, 'T': 16, 'V': 17, 'W': 18, 'Y': 19}\n",
      "getting AA sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "viruses no files 0\n",
      "X_training:Extacting  AA  of length  1  from   42 sequences\n",
      "(42, 20)\n",
      "X_test: Extacting  AA  of length  1  from   14 sequences\n",
      "(14, 20)\n",
      "(42, 20) (14, 20) (42,) (14,)\n",
      "{'AUC': 0.729, 'N in class': 28, 'Features': 'AA', 'k': 1}\n",
      "X_training:Extacting  AA  of length  2  from   42 sequences\n",
      "(42, 400)\n",
      "X_test: Extacting  AA  of length  2  from   14 sequences\n",
      "(14, 400)\n",
      "(42, 400) (14, 400) (42,) (14,)\n",
      "{'AUC': 0.75, 'N in class': 28, 'Features': 'AA', 'k': 2}\n",
      "X_training:Extacting  AA  of length  3  from   42 sequences\n",
      "(42, 8000)\n",
      "X_test: Extacting  AA  of length  3  from   14 sequences\n",
      "(14, 8000)\n",
      "(42, 8000) (14, 8000) (42,) (14,)\n",
      "{'AUC': 0.833, 'N in class': 28, 'Features': 'AA', 'k': 3}\n",
      "X_training:Extacting  AA  of length  4  from   42 sequences\n",
      "(42, 160000)\n",
      "X_test: Extacting  AA  of length  4  from   14 sequences\n",
      "(14, 160000)\n",
      "(42, 160000) (14, 160000) (42,) (14,)\n",
      "{'AUC': 0.875, 'N in class': 28, 'Features': 'AA', 'k': 4}\n",
      "2 [1, 2, 3, 4, 5, 6] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 7, 'C': 0, 'A': 1, 'G': 1, 'V': 1, 'I': 2, 'L': 2, 'F': 2, 'P': 2, 'M': 3, 'S': 3, 'T': 3, 'Y': 3, 'H': 4, 'N': 4, 'Q': 4, 'W': 4, 'R': 5, 'K': 5, 'D': 6, 'E': 6}\n",
      "getting PC sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "viruses no files 0\n",
      "X_training:Extacting  PC  of length  1  from   42 sequences\n",
      "(42, 7)\n",
      "X_test: Extacting  PC  of length  1  from   14 sequences\n",
      "(14, 7)\n",
      "(42, 7) (14, 7) (42,) (14,)\n",
      "{'AUC': 0.646, 'N in class': 28, 'Features': 'PC', 'k': 1}\n",
      "X_training:Extacting  PC  of length  2  from   42 sequences\n",
      "(42, 49)\n",
      "X_test: Extacting  PC  of length  2  from   14 sequences\n",
      "(14, 49)\n",
      "(42, 49) (14, 49) (42,) (14,)\n",
      "{'AUC': 0.75, 'N in class': 28, 'Features': 'PC', 'k': 2}\n",
      "X_training:Extacting  PC  of length  3  from   42 sequences\n",
      "(42, 343)\n",
      "X_test: Extacting  PC  of length  3  from   14 sequences\n",
      "(14, 343)\n",
      "(42, 343) (14, 343) (42,) (14,)\n",
      "{'AUC': 0.896, 'N in class': 28, 'Features': 'PC', 'k': 3}\n",
      "X_training:Extacting  PC  of length  4  from   42 sequences\n",
      "(42, 2401)\n",
      "X_test: Extacting  PC  of length  4  from   14 sequences\n",
      "(14, 2401)\n",
      "(42, 2401) (14, 2401) (42,) (14,)\n",
      "{'AUC': 0.854, 'N in class': 28, 'Features': 'PC', 'k': 4}\n",
      "X_training:Extacting  PC  of length  5  from   42 sequences\n",
      "(42, 16807)\n",
      "X_test: Extacting  PC  of length  5  from   14 sequences\n",
      "(14, 16807)\n",
      "(42, 16807) (14, 16807) (42,) (14,)\n",
      "{'AUC': 0.833, 'N in class': 28, 'Features': 'PC', 'k': 5}\n",
      "X_training:Extacting  PC  of length  6  from   42 sequences\n",
      "(42, 117649)\n",
      "X_test: Extacting  PC  of length  6  from   14 sequences\n",
      "(14, 117649)\n",
      "(42, 117649) (14, 117649) (42,) (14,)\n",
      "{'AUC': 0.896, 'N in class': 28, 'Features': 'PC', 'k': 6}\n",
      "3 [0] /home4/youn01f/Desktop/workspace/newData/pfs {}\n",
      "getting Domains sequences from /home4/youn01f/Desktop/workspace/newData/pfs\n",
      "viruses no files 1\n",
      "X_training:Extacting  Domains  of length  0  from   42 sequences\n",
      "(42, 103)\n",
      "X_test: Extacting  Domains  of length  0  from   13 sequences\n",
      "(13, 103)\n",
      "(42, 103) (13, 103) (42,) (13,)\n",
      "{'AUC': 0.8, 'N in class': 28, 'Features': 'Domains', 'k': 0}\n",
      " size of training 224 and test set   58\n",
      "Bovidae family Mammalia class RNA\n",
      "0 [1, 2, 3, 4, 5, 6, 7, 8, 9] /home4/youn01f/Desktop/workspace/newData/fasta {'mod': 4, 'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
      "getting DNA sequences from /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "viruses no files 6\n",
      "X_training:Extacting  DNA  of length  1  from   219 sequences\n",
      "(219, 4)\n",
      "X_test: Extacting  DNA  of length  1  from   57 sequences\n",
      "(57, 4)\n",
      "(219, 4) (57, 4) (219,) (57,)\n",
      "{'AUC': 0.571, 'N in class': 141, 'Features': 'DNA', 'k': 1}\n",
      "X_training:Extacting  DNA  of length  2  from   219 sequences\n",
      "(219, 16)\n",
      "X_test: Extacting  DNA  of length  2  from   57 sequences\n",
      "(57, 16)\n",
      "(219, 16) (57, 16) (219,) (57,)\n",
      "{'AUC': 0.748, 'N in class': 141, 'Features': 'DNA', 'k': 2}\n",
      "X_training:Extacting  DNA  of length  3  from   219 sequences\n",
      "(219, 64)\n",
      "X_test: Extacting  DNA  of length  3  from   57 sequences\n",
      "(57, 64)\n",
      "(219, 64) (57, 64) (219,) (57,)\n",
      "{'AUC': 0.77, 'N in class': 141, 'Features': 'DNA', 'k': 3}\n",
      "X_training:Extacting  DNA  of length  4  from   219 sequences\n",
      "(219, 256)\n",
      "X_test: Extacting  DNA  of length  4  from   57 sequences\n",
      "(57, 256)\n",
      "(219, 256) (57, 256) (219,) (57,)\n",
      "{'AUC': 0.792, 'N in class': 141, 'Features': 'DNA', 'k': 4}\n",
      "X_training:Extacting  DNA  of length  5  from   219 sequences\n",
      "(219, 1024)\n",
      "X_test: Extacting  DNA  of length  5  from   57 sequences\n",
      "(57, 1024)\n",
      "(219, 1024) (57, 1024) (219,) (57,)\n",
      "{'AUC': 0.878, 'N in class': 141, 'Features': 'DNA', 'k': 5}\n",
      "X_training:Extacting  DNA  of length  6  from   219 sequences\n",
      "(219, 4096)\n",
      "X_test: Extacting  DNA  of length  6  from   57 sequences\n",
      "(57, 4096)\n",
      "(219, 4096) (57, 4096) (219,) (57,)\n",
      "{'AUC': 0.913, 'N in class': 141, 'Features': 'DNA', 'k': 6}\n",
      "X_training:Extacting  DNA  of length  7  from   219 sequences\n"
     ]
    }
   ],
   "source": [
    "for subset in subsets:\n",
    "    class_data,n = fp.get_class_lists(subset,viruses,hosts)\n",
    "    datasets = fp.split_data(viruses,class_data,n)\n",
    "    (label,label_tax,pool,pool_tax,baltimore) = subset\n",
    "    print  (label,label_tax,pool,pool_tax,baltimore)\n",
    "   \n",
    "    for feature in feature_list:\n",
    "        \n",
    "        index = features.index(feature)\n",
    "        kmers = kmer_lists[index]\n",
    "        filepath = filepaths[index]\n",
    "        symbol_dict = symbol_dicts[index]\n",
    "        ext = file_exts[index]\n",
    "        print (index,kmers,filepath,symbol_dict)\n",
    "\n",
    "        if features != 'PC': #PC same sequence as for AA\n",
    "            filepath = filepaths[index]\n",
    "            print ( f'getting {feature} sequences from {filepath}')\n",
    "            sequences = get_sequences(filepath,datasets,ext)\n",
    "\n",
    "        for k in kmers: \n",
    "            x_train,x_test,y_train,y_test = get_feature_matrices(sequences,datasets,k,symbol_dict)\n",
    "            print (np.shape(x_train), np.shape(x_test),np.shape(y_train),np.shape(y_test))\n",
    "            #print (type(x_train), type(x_test),type(y_train),type(y_test))\n",
    "            results = fp.test_prediction(x_train,x_test,y_train,y_test)\n",
    "            results.update ({'N in class': n, 'Features':feature, 'k':k})\n",
    "            print(results)\n",
    "            fp.results2CSV (results,subset, results_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test one subset on on feature type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(datasets['training'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_list = ['DNA','AA','PC','Domains']\n",
    "features = ['AA']\n",
    "for feature in features:\n",
    "    index = feature_list.index(feature)\n",
    "    kmers = kmer_lists[index]\n",
    "    filepath = filepaths[index]\n",
    "    symbol_dict = symbol_dicts[index]\n",
    "    ext = file_exts[index]\n",
    "    print (index,kmers,filepath,symbol_dict)\n",
    "    \n",
    "    if features != 'PC': #PC same sequence as for AA\n",
    "        filepath = filepaths[index]\n",
    "        print ( f'getting {feature} sequences from {filepath}')\n",
    "        sequences = get_sequences(filepath,datasets,ext)\n",
    "    \n",
    "    for k in kmers: \n",
    "        start_time = timeit.default_timer()\n",
    "        x_train,x_test,y_train,y_test = get_feature_matrices(sequences,datasets,k,symbol_dict)\n",
    "        print ('timeit: ',timeit.default_timer() - start_time)\n",
    "        print (np.shape(x_train), np.shape(x_test),np.shape(y_train),np.shape(y_test))\n",
    "        results = fp.test_prediction(x_train,x_test,y_train,y_test)\n",
    "        results.update ({'N in class': n, 'Features':feature, 'k':k})\n",
    "        print(results)\n",
    "        fp.results2CSV (results,subset, results_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
