{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "import os\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import random\n",
    "import numpy as np\n",
    "from collections import defaultdict, Counter\n",
    "import csv\n",
    "\n",
    "import sys\n",
    "lib_dir = '../mylibs'\n",
    "if lib_dir not in sys.path:\n",
    "    sys.path.append(lib_dir)\n",
    "    \n",
    "import vhdb as vhdb\n",
    "import features_predict as fp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "filein = '../inputs/VHDB_25_1_2019.p'\n",
    "with open(filein, 'rb') as f:\n",
    "    V_H = pickle.load( f)\n",
    "hosts = V_H.hosts\n",
    "viruses = V_H.viruses\n",
    "subspecies =V_H.subspecies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bact_holdout_results.csv\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Input file\n",
    "#subsets = [(label,taxlevel,pool,taxlevelpool),...]\n",
    "subsetfile = 'Bact_holdout.csv'\n",
    "baltimore = 'all'\n",
    "vhdbfile = '../inputs/VHDB_25_1_2019.p'\n",
    "\n",
    "\n",
    "# Everthing in the same order as the feature_function list\n",
    "# a list of the functions to be used create the feature matrix\n",
    "\n",
    "feature_functions = [\n",
    "                    fp.DNA_features,fp.AA_features,fp.PC_features,\n",
    "                    fp.Domain_features]\n",
    "\n",
    "# filepath for the fasta, faa and domain files\n",
    "filepaths = [\n",
    "                '/home4/youn01f/Desktop/workspace/newData/fasta',\n",
    "                '/home4/youn01f/Desktop/workspace/newData/faa','/home4/youn01f/Desktop/workspace/newData/faa',\n",
    "                '/home4/youn01f/Desktop/workspace/newData/pfs',]\n",
    "\n",
    "# A list of kmers for each feature set to be tested\n",
    "kmer_lists = [\n",
    "              [3,6,9], # dna \n",
    "              [3,4], # aa\n",
    "              [5,6] ,    #pc\n",
    "              [1]] #domains\n",
    "\n",
    "\n",
    "# symbol dictionaries \n",
    "na_dict = {'mod':4,'a':0,'c':1,'g':2,'t':3}\n",
    "aa_dict = {'mod':20 ,'A':0,'C':1,'D':2,'E':3,'F':4,'G':5,'H':6,'I':7,'K':8,'L':9,'M':10,'N':11,\n",
    "              'P':12,'Q':13,'R':14,'S':15,'T':16,'V':17,'W':18,'Y':19}\n",
    "pc_dict = {'mod':7, 'C':0,\n",
    "            'A':1,'G':1,'V':1,\n",
    "           'I':2,'L':2,'F':2,'P':2,\n",
    "           'M':3,'S':3,'T':3,'Y':3,\n",
    "           'H':4,'N':4,'Q':4,'W':4,\n",
    "           'R':5,'K':5,\n",
    "           'D':6,'E':6}\n",
    "symbol_dicts = [\n",
    "                na_dict,aa_dict,pc_dict, \n",
    "                {}]\n",
    "\n",
    "classifiers = ['SVM_lin']\n",
    "\n",
    "# Output file for the results\n",
    "dataset = subsetfile.split('.')[0]\n",
    "results_file =  f'{dataset}_results.csv'\n",
    "model = 1\n",
    "print (results_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Bacteroidetes', 'phylum', 'Bacteria', 'kingdom', 'DNA', 'Siphoviridae'),\n",
       " ('Actinobacteria', 'class', 'Bacteria', 'kingdom', 'DNA', 'Siphoviridae'),\n",
       " ('Firmicutes', 'phylum', 'Bacteria', 'kingdom', 'DNA', 'Siphoviridae')]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "label_info = pd.read_csv(subsetfile)\n",
    "subsets = label_info.apply(tuple, axis=1).tolist()\n",
    "(subsets[:3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alter get_class_lists this for holdout virus group\n",
    "use training with holdout data use fp.get_class_losts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_class_lists_ho(subset,viruses,hosts):\n",
    "    \n",
    "    (label,label_tax,pool,pool_tax,balt,v_holdout) = subset\n",
    "    data_lists = {'training':[[],[]],'test':[[],[]]}\n",
    "    # Get a list of all the viruses in the labelled class and the rest of the pool\n",
    "    \n",
    "    \n",
    "    baltimore = 'NA'  # all slasses (except satallites)\n",
    "   \n",
    "    for v, vd in viruses.items() :\n",
    "        if  baltimore in vd['class']: \n",
    "            \n",
    "            if  vd['family']!= v_holdout:\n",
    "                ds_name ='training'\n",
    "            else:\n",
    "                ds_name = 'test'\n",
    "            for h in vd['hosts']:\n",
    "                if hosts[h][pool_tax.lower()] == pool:\n",
    "                    if hosts[h][label_tax.lower()] ==label:\n",
    "                        \n",
    "                        data_lists[ds_name][0].append ((v,label))\n",
    "                    else:\n",
    "                        data_lists[ds_name][1].append ((v,'Other'))\n",
    "                \n",
    "           \n",
    "   # print (f'training - positive {len(data_lists[\"training\"][0])} negative {len(data_lists[\"training\"][1])})') \n",
    "   # print (f'test - positive {len(data_lists[\"test\"][0])} negative {len(data_lists[\"test\"][1])}')\n",
    "\n",
    "    # Get a random sample for each class of size n , the size of smallest class                    \n",
    "   \n",
    "   \n",
    " # Keep all the viruses in the holdout set as test data. they will not be same size   \n",
    "    datasets = {'training':{},'test':{}}\n",
    "    nmax = {'training':400,'test':50}\n",
    "    for k,data_list in data_lists.items():\n",
    "        datas =[]\n",
    "        n = min (len(data_list[0]),len(data_list[1]),nmax[k])\n",
    "\n",
    "        #print (f'number of v {k} {n}')\n",
    "        #n = 6 #test\n",
    "        # randomly choose n viruses from both classes\n",
    "        for clss in data_list: \n",
    "            sample  = random.sample(range(len(clss)), n)\n",
    "            for s in sample:\n",
    "                datas.append (clss[s])\n",
    "            \n",
    "       # shuffle together\n",
    "        random.shuffle(datas)\n",
    "        for v,l in datas:\n",
    "            datasets [k][v]= {'label':l, 'refseqs':viruses[v]['refseqs']}\n",
    "    ntrain = (len(datasets['training']))\n",
    "    ntest = (len(datasets['test']))\n",
    "    #print(f' ho training v {ntrain} , test v {ntest}')\n",
    "    return datasets,ntrain,ntest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Adapt get_class_list to give infomation about number of holdout viruses in test set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_class_lists(subset,viruses,hosts):\n",
    "# returns two list of viruses one for each  positive class and negative class    \n",
    "    (label,label_tax,pool,pool_tax,baltimore,v_ho) = subset\n",
    "    \n",
    "    if baltimore =='all':\n",
    "        baltimore = 'NA'  # all classes  contain either 'DNA' or 'RNA'\n",
    "    \n",
    "    pos_neglists =  [[],[]]\n",
    "    # Get 2 lists of all the viruses in the labelled class and the rest of the pool\n",
    "\n",
    "    for v, vd in viruses.items() :\n",
    "        if baltimore in vd['class']:          \n",
    "            for h in vd['hosts']:\n",
    "                if hosts[h][pool_tax.lower()] == pool:\n",
    "                    if hosts[h][label_tax.lower()] ==label:\n",
    "                        pos_neglists[0].append ((v,label))\n",
    "                    else:\n",
    "                         pos_neglists[1].append((v,'Other'))\n",
    "    #print('all pos and neg v', len(pos_neglists[0]),len(pos_neglists[1]))                    \n",
    "    # Get a random sample for each class of size n , the size of smallest class \n",
    "    #restrict to 400\n",
    "    n = min(len(pos_neglists[0]),len(pos_neglists[1]),400)\n",
    "   # n=6 # test\n",
    "    random.seed(10)\n",
    "    datas =  []\n",
    "    for clss in pos_neglists:\n",
    "        data = []\n",
    "        sample  = random.sample(range(len(clss)), n)\n",
    "        for s in sample:\n",
    "            data.append (clss[s])\n",
    "        datas.append(data)\n",
    "    #print('pos and neg lists',len(datas[0]),len(datas[1]))   \n",
    "    #split into training and test sets     \n",
    "    split = 0.75 \n",
    "    tntlist =  [[],[]]  #2 lists for training and test viruses\n",
    "    \n",
    "    for viruslist in datas:\n",
    "        for i in range(n):\n",
    "            if random.random() < split:\n",
    "                tntlist[0].append(viruslist[i]) \n",
    "            else:\n",
    "                tntlist[1].append(viruslist[i]) \n",
    "   \n",
    "    datadicts = {'training':{},'test':{}}\n",
    "    for i,k in enumerate(datadicts):\n",
    "        random.shuffle(tntlist[i])\n",
    "\n",
    "        for v,l in tntlist[i]:\n",
    "            datadicts [k][v]= {'label':l, 'refseqs':viruses[v]['refseqs']}\n",
    "       \n",
    "        ntrain  =len(datadicts['training'])\n",
    "        ntest = len(datadicts['test'])\n",
    "        ho_test =[]\n",
    "        for v in datadicts['test']:\n",
    "            if viruses[v]['family']== v_ho:\n",
    "                ho_test.append(v)\n",
    "        n_ho_test = len (ho_test)\n",
    "   # print (f'all- size of training and test set {ntrain},  {ntest}')\n",
    "    \n",
    "    return (datadicts,ntrain,n_ho_test)\n",
    "    \n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "model =0\n",
    "def test_dataset(subset,datasets,n): \n",
    "    (label,label_tax,pool,pool_tax,balt,v_holdout,t_g) = subset\n",
    "    global model\n",
    "    print(model)\n",
    "    for feature_func in feature_functions:\n",
    "        index = feature_functions.index(feature_func)\n",
    "        filepath = filepaths[index]\n",
    "        kmers = kmer_lists[index]\n",
    "        symbol_dict = symbol_dicts[index]\n",
    "        features = (feature_func.__name__).split('_')[0]\n",
    "        for k in kmers:\n",
    "\n",
    "            model += 1\n",
    "        # get feature matrix , test prediction and write results to CSV file\n",
    "            print (label,label_tax,feature_func.__name__ ,k,filepath)\n",
    "            (X,X_test), v_index = feature_func (datasets,k,symbol_dict,filepath)\n",
    "            y, y_test = fp.get_labels(label,v_index, datasets)\n",
    "            print ( 'X_train,Y_train',np.shape(X), np.shape(y))\n",
    "            results, pred_probs = fp.test_prediction2(X,y,X_test, y_test)\n",
    "            # Add info for results\n",
    "            results.update ({ 'N in class': n, 'Features':features, 'k': k})\n",
    "            \n",
    "            vs = []\n",
    "            v_fam = []\n",
    "            ys = []\n",
    "            probs = []\n",
    "            for  v,y,prob in zip(v_index[1],y_test,pred_probs):\n",
    "                #print(v, hosts[viruses[v]['hosts'][0]][label_tax],y,prob)\n",
    "                #print (viruses[v]['family'])\n",
    "                if viruses[v]['family']== v_holdout:\n",
    "                    vs.append(v)\n",
    "                    v_fam.append(viruses[v]['family'])\n",
    "                    ys.append(y)\n",
    "                    probs.append(prob)\n",
    "            print('length holdout data', len(vs),len(ys),len(probs))\n",
    "            ss = (model,*subset,features,k)\n",
    "            probs2csv(vs,v_fam,ys, probs,ss, dataset)\n",
    "            results.update({'positive label':label ,'label tax group': label_tax,'pool label':pool,\n",
    "                            'pool tax group':pool_tax, 'Baltimore':balt,\n",
    "                            'virus holdout group':v_holdout,'training group':t_g,})\n",
    "            print(results)\n",
    "            results2CSV (results, results_file)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def results2CSV(results, csvfile):\n",
    "    fieldnames = ['positive label','label tax group','pool label','pool tax group',\n",
    "                          'Baltimore','virus holdout group','training group', 'N in class' , \n",
    "                          'Features','k','AUC','accuracy', 'specificity','sensitivity', 'TN/FP/FN/TP' ]\n",
    "    if os.path.isfile(csvfile):\n",
    "        with open(csvfile, 'a') as csvfile:\n",
    "            #fieldnames = ['positive label','label tax group','pool label','pool tax group','Baltimore','virus holdout group','training group', 'N in class' , 'Features','k','AUC','accuracy', 'specificity','sensitivity', 'TN/FP/FN/TP' ]\n",
    "            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)\n",
    "            writer.writerow(results) \n",
    "    else:\n",
    "        with open(csvfile, 'a') as csvfile:\n",
    "#             fieldnames = ['positive label','label tax group','pool label','pool tax group',\n",
    "#                           'Baltimore','virus holdout group','training group', 'N in class' , \n",
    "#                           'Features','k','AUC','accuracy', 'specificity','sensitivity', 'TN/FP/FN/TP' ]\n",
    "            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)\n",
    "            writer.writeheader()\n",
    "            writer.writerow(results)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def probs2csv (v_index, v_fam, y_test, pred_probs,subset, dataset):\n",
    "    \n",
    "    csvfile =  f'{dataset}_probs.csv'\n",
    "    fieldnames = ['model','label','label_tax','pool','pool_tax','balt','v_holdout',\n",
    "    'training group','features','k', 'Virus','virus family', 'true label', 'predict prob']\n",
    "    \n",
    "    if os.path.isfile(csvfile):\n",
    "        with open(csvfile, 'a') as csvfile:\n",
    "            writer = csv.writer(csvfile)\n",
    "            for virus,f,y,y_pred in zip(v_index,v_fam,y_test, pred_probs):\n",
    "                row = (*subset,virus,f,y,y_pred )\n",
    "                writer.writerow(row) \n",
    "                \n",
    "    else:\n",
    "        with open(csvfile, 'w') as csvfile:\n",
    "            writer = csv.writer(csvfile)\n",
    "            writer.writerow(fieldnames)\n",
    "            for virus,f,y,y_pred in zip(v_index,v_fam, y_test, pred_probs):\n",
    "                row = (*subset,virus,f,y,y_pred )\n",
    "                writer.writerow(row) \n",
    "    \n",
    "    \n",
    "    return"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get AUC for each feature set for each data set. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "Bacteroidetes phylum DNA_features 3 /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "start of making  training X,63, (63, 64)\n",
      "63 (63, 64)\n",
      "start of making  test X,51, (51, 64)\n",
      "51 (51, 64)\n",
      "labels y 0,63\n",
      "labels y 1,51\n",
      "X_train,Y_train (63, 64) (63,)\n",
      "length holdout data 51 51 51\n",
      "{'AUC': 0.95, 'accuracy': 0.804, 'specificity': 0.69, 'sensitivity': 0.955, 'TN/FP/FN/TP': array([20,  9,  1, 21]), 'N in class': 63, 'Features': 'DNA', 'k': 3, 'positive label': 'Bacteroidetes', 'label tax group': 'phylum', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "Bacteroidetes phylum DNA_features 6 /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "start of making  training X,63, (63, 4096)\n",
      "63 (63, 4096)\n",
      "start of making  test X,51, (51, 4096)\n",
      "51 (51, 4096)\n",
      "labels y 0,63\n",
      "labels y 1,51\n",
      "X_train,Y_train (63, 4096) (63,)\n",
      "length holdout data 51 51 51\n",
      "{'AUC': 0.989, 'accuracy': 0.941, 'specificity': 1.0, 'sensitivity': 0.864, 'TN/FP/FN/TP': array([29,  0,  3, 19]), 'N in class': 63, 'Features': 'DNA', 'k': 6, 'positive label': 'Bacteroidetes', 'label tax group': 'phylum', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "Bacteroidetes phylum DNA_features 9 /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "start of making  training X,63, (63, 262144)\n",
      "63 (63, 262144)\n",
      "start of making  test X,51, (51, 262144)\n",
      "51 (51, 262144)\n",
      "labels y 0,63\n",
      "labels y 1,51\n",
      "X_train,Y_train (63, 262144) (63,)\n",
      "length holdout data 51 51 51\n",
      "{'AUC': 0.98, 'accuracy': 0.588, 'specificity': 1.0, 'sensitivity': 0.045, 'TN/FP/FN/TP': array([29,  0, 21,  1]), 'N in class': 63, 'Features': 'DNA', 'k': 9, 'positive label': 'Bacteroidetes', 'label tax group': 'phylum', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "Bacteroidetes phylum AA_features 3 /home4/youn01f/Desktop/workspace/newData/faa\n",
      "start of making  training X,63, (63, 8000)\n",
      "63 (63, 8000)\n",
      "start of making  test X,51, (51, 8000)\n",
      "51 (51, 8000)\n",
      "labels y 0,63\n",
      "labels y 1,51\n",
      "X_train,Y_train (63, 8000) (63,)\n",
      "length holdout data 51 51 51\n",
      "{'AUC': 0.975, 'accuracy': 0.569, 'specificity': 1.0, 'sensitivity': 0.0, 'TN/FP/FN/TP': array([29,  0, 22,  0]), 'N in class': 63, 'Features': 'AA', 'k': 3, 'positive label': 'Bacteroidetes', 'label tax group': 'phylum', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "Bacteroidetes phylum AA_features 4 /home4/youn01f/Desktop/workspace/newData/faa\n",
      "start of making  training X,63, (63, 160000)\n",
      "63 (63, 160000)\n",
      "start of making  test X,51, (51, 160000)\n",
      "51 (51, 160000)\n",
      "labels y 0,63\n",
      "labels y 1,51\n",
      "X_train,Y_train (63, 160000) (63,)\n",
      "length holdout data 51 51 51\n",
      "{'AUC': 0.917, 'accuracy': 0.569, 'specificity': 1.0, 'sensitivity': 0.0, 'TN/FP/FN/TP': array([29,  0, 22,  0]), 'N in class': 63, 'Features': 'AA', 'k': 4, 'positive label': 'Bacteroidetes', 'label tax group': 'phylum', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "Bacteroidetes phylum PC_features 5 /home4/youn01f/Desktop/workspace/newData/faa\n",
      "start of making  training X,63, (63, 16807)\n",
      "63 (63, 16807)\n",
      "start of making  test X,51, (51, 16807)\n",
      "51 (51, 16807)\n",
      "labels y 0,63\n",
      "labels y 1,51\n",
      "X_train,Y_train (63, 16807) (63,)\n",
      "length holdout data 51 51 51\n",
      "{'AUC': 0.862, 'accuracy': 0.569, 'specificity': 1.0, 'sensitivity': 0.0, 'TN/FP/FN/TP': array([29,  0, 22,  0]), 'N in class': 63, 'Features': 'PC', 'k': 5, 'positive label': 'Bacteroidetes', 'label tax group': 'phylum', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "Bacteroidetes phylum PC_features 6 /home4/youn01f/Desktop/workspace/newData/faa\n",
      "start of making  training X,63, (63, 117649)\n",
      "63 (63, 117649)\n",
      "start of making  test X,51, (51, 117649)\n",
      "51 (51, 117649)\n",
      "labels y 0,63\n",
      "labels y 1,51\n",
      "X_train,Y_train (63, 117649) (63,)\n",
      "length holdout data 51 51 51\n",
      "{'AUC': 0.868, 'accuracy': 0.569, 'specificity': 1.0, 'sensitivity': 0.0, 'TN/FP/FN/TP': array([29,  0, 22,  0]), 'N in class': 63, 'Features': 'PC', 'k': 6, 'positive label': 'Bacteroidetes', 'label tax group': 'phylum', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "Bacteroidetes phylum Domain_features 1 /home4/youn01f/Desktop/workspace/newData/pfs\n",
      "There are 114 viruses with domains, 0 viruses missing domain info\n",
      "63 (63, 766)\n",
      "51 (51, 766)\n",
      "labels y 0,63\n",
      "labels y 1,51\n",
      "X_train,Y_train (63, 766) (63,)\n",
      "length holdout data 51 51 51\n",
      "{'AUC': 0.458, 'accuracy': 0.588, 'specificity': 0.897, 'sensitivity': 0.182, 'TN/FP/FN/TP': array([26,  3, 18,  4]), 'N in class': 63, 'Features': 'Domain', 'k': 1, 'positive label': 'Bacteroidetes', 'label tax group': 'phylum', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "8\n",
      "Bacteroidetes phylum DNA_features 3 /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "start of making  training X,89, (89, 64)\n",
      "89 (89, 4096)\n",
      "start of making  test X,30, (30, 4096)\n",
      "30 (30, 4096)\n",
      "labels y 0,89\n",
      "labels y 1,30\n",
      "X_train,Y_train (89, 4096) (89,)\n",
      "length holdout data 17 17 17\n",
      "{'AUC': 1.0, 'accuracy': 0.933, 'specificity': 0.875, 'sensitivity': 1.0, 'TN/FP/FN/TP': array([14,  2,  0, 14]), 'N in class': 89, 'Features': 'DNA', 'k': 6, 'positive label': 'Bacteroidetes', 'label tax group': 'phylum', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'All'}\n",
      "Bacteroidetes phylum DNA_features 9 /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "start of making  training X,89, (89, 262144)\n",
      "89 (89, 262144)\n",
      "start of making  test X,30, (30, 262144)\n",
      "30 (30, 262144)\n",
      "labels y 0,89\n",
      "labels y 1,30\n",
      "X_train,Y_train (89, 262144) (89,)\n",
      "length holdout data 17 17 17\n",
      "{'AUC': 1.0, 'accuracy': 0.967, 'specificity': 0.938, 'sensitivity': 1.0, 'TN/FP/FN/TP': array([15,  1,  0, 14]), 'N in class': 89, 'Features': 'DNA', 'k': 9, 'positive label': 'Bacteroidetes', 'label tax group': 'phylum', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'All'}\n",
      "Bacteroidetes phylum AA_features 3 /home4/youn01f/Desktop/workspace/newData/faa\n",
      "start of making  training X,89, (89, 8000)\n",
      "89 (89, 8000)\n",
      "start of making  test X,30, (30, 8000)\n",
      "29 (29, 8000)\n",
      "labels y 0,89\n",
      "labels y 1,29\n",
      "X_train,Y_train (89, 8000) (89,)\n",
      "length holdout data 17 17 17\n",
      "{'AUC': 1.0, 'accuracy': 1.0, 'specificity': 1.0, 'sensitivity': 1.0, 'TN/FP/FN/TP': array([15,  0,  0, 14]), 'N in class': 89, 'Features': 'AA', 'k': 3, 'positive label': 'Bacteroidetes', 'label tax group': 'phylum', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'All'}\n",
      "Bacteroidetes phylum AA_features 4 /home4/youn01f/Desktop/workspace/newData/faa\n",
      "start of making  training X,89, (89, 160000)\n",
      "89 (89, 160000)\n",
      "start of making  test X,30, (30, 160000)\n",
      "29 (29, 160000)\n",
      "labels y 0,89\n",
      "labels y 1,29\n",
      "X_train,Y_train (89, 160000) (89,)\n",
      "length holdout data 17 17 17\n",
      "{'AUC': 1.0, 'accuracy': 0.897, 'specificity': 1.0, 'sensitivity': 0.786, 'TN/FP/FN/TP': array([15,  0,  3, 11]), 'N in class': 89, 'Features': 'AA', 'k': 4, 'positive label': 'Bacteroidetes', 'label tax group': 'phylum', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'All'}\n",
      "Bacteroidetes phylum PC_features 5 /home4/youn01f/Desktop/workspace/newData/faa\n",
      "start of making  training X,89, (89, 16807)\n",
      "86 (86, 64)\n",
      "start of making  test X,98, (98, 64)\n",
      "98 (98, 64)\n",
      "labels y 0,86\n",
      "labels y 1,98\n",
      "X_train,Y_train (86, 64) (86,)\n",
      "length holdout data 98 98 98\n",
      "{'AUC': 0.974, 'accuracy': 0.929, 'specificity': 0.94, 'sensitivity': 0.917, 'TN/FP/FN/TP': array([47,  3,  4, 44]), 'N in class': 86, 'Features': 'DNA', 'k': 3, 'positive label': 'Actinobacteria', 'label tax group': 'class', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "Actinobacteria class DNA_features 6 /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "start of making  training X,86, (86, 4096)\n",
      "86 (86, 4096)\n",
      "start of making  test X,98, (98, 4096)\n",
      "98 (98, 4096)\n",
      "labels y 0,86\n",
      "labels y 1,98\n",
      "X_train,Y_train (86, 4096) (86,)\n",
      "length holdout data 98 98 98\n",
      "{'AUC': 0.982, 'accuracy': 0.857, 'specificity': 0.94, 'sensitivity': 0.771, 'TN/FP/FN/TP': array([47,  3, 11, 37]), 'N in class': 86, 'Features': 'DNA', 'k': 6, 'positive label': 'Actinobacteria', 'label tax group': 'class', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "Actinobacteria class DNA_features 9 /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "start of making  training X,86, (86, 262144)\n",
      "86 (86, 262144)\n",
      "start of making  test X,98, (98, 262144)\n",
      "98 (98, 262144)\n",
      "labels y 0,86\n",
      "labels y 1,98\n",
      "X_train,Y_train (86, 262144) (86,)\n",
      "length holdout data 98 98 98\n",
      "{'AUC': 0.935, 'accuracy': 0.816, 'specificity': 0.88, 'sensitivity': 0.75, 'TN/FP/FN/TP': array([44,  6, 12, 36]), 'N in class': 86, 'Features': 'DNA', 'k': 9, 'positive label': 'Actinobacteria', 'label tax group': 'class', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "Actinobacteria class AA_features 3 /home4/youn01f/Desktop/workspace/newData/faa\n",
      "start of making  training X,86, (86, 8000)\n",
      "86 (86, 8000)\n",
      "start of making  test X,98, (98, 8000)\n",
      "98 (98, 8000)\n",
      "labels y 0,86\n",
      "labels y 1,98\n",
      "X_train,Y_train (86, 8000) (86,)\n",
      "length holdout data 98 98 98\n",
      "{'AUC': 0.979, 'accuracy': 0.929, 'specificity': 0.9, 'sensitivity': 0.958, 'TN/FP/FN/TP': array([45,  5,  2, 46]), 'N in class': 86, 'Features': 'AA', 'k': 3, 'positive label': 'Actinobacteria', 'label tax group': 'class', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "Actinobacteria class AA_features 4 /home4/youn01f/Desktop/workspace/newData/faa\n",
      "start of making  training X,86, (86, 160000)\n",
      "86 (86, 160000)\n",
      "start of making  test X,98, (98, 160000)\n",
      "98 (98, 160000)\n",
      "labels y 0,86\n",
      "labels y 1,98\n",
      "X_train,Y_train (86, 160000) (86,)\n",
      "length holdout data 98 98 98\n",
      "{'AUC': 0.96, 'accuracy': 0.51, 'specificity': 1.0, 'sensitivity': 0.0, 'TN/FP/FN/TP': array([50,  0, 48,  0]), 'N in class': 86, 'Features': 'AA', 'k': 4, 'positive label': 'Actinobacteria', 'label tax group': 'class', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "Actinobacteria class PC_features 5 /home4/youn01f/Desktop/workspace/newData/faa\n",
      "start of making  training X,86, (86, 16807)\n",
      "86 (86, 16807)\n",
      "start of making  test X,98, (98, 16807)\n",
      "98 (98, 16807)\n",
      "labels y 0,86\n",
      "labels y 1,98\n",
      "X_train,Y_train (86, 16807) (86,)\n",
      "length holdout data 98 98 98\n",
      "{'AUC': 0.927, 'accuracy': 0.602, 'specificity': 1.0, 'sensitivity': 0.188, 'TN/FP/FN/TP': array([50,  0, 39,  9]), 'N in class': 86, 'Features': 'PC', 'k': 5, 'positive label': 'Actinobacteria', 'label tax group': 'class', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "Actinobacteria class PC_features 6 /home4/youn01f/Desktop/workspace/newData/faa\n",
      "start of making  training X,86, (86, 117649)\n",
      "86 (86, 117649)\n",
      "start of making  test X,98, (98, 117649)\n",
      "98 (98, 117649)\n",
      "labels y 0,86\n",
      "labels y 1,98\n",
      "X_train,Y_train (86, 117649) (86,)\n",
      "length holdout data 98 98 98\n",
      "{'AUC': 0.93, 'accuracy': 0.51, 'specificity': 1.0, 'sensitivity': 0.0, 'TN/FP/FN/TP': array([50,  0, 48,  0]), 'N in class': 86, 'Features': 'PC', 'k': 6, 'positive label': 'Actinobacteria', 'label tax group': 'class', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "Actinobacteria class Domain_features 1 /home4/youn01f/Desktop/workspace/newData/pfs\n",
      "There are 183 viruses with domains, 1 viruses missing domain info\n",
      "85 (85, 971)\n",
      "98 (98, 971)\n",
      "labels y 0,85\n",
      "labels y 1,98\n",
      "X_train,Y_train (85, 971) (85,)\n",
      "length holdout data 98 98 98\n",
      "{'AUC': 0.836, 'accuracy': 0.612, 'specificity': 1.0, 'sensitivity': 0.208, 'TN/FP/FN/TP': array([50,  0, 38, 10]), 'N in class': 86, 'Features': 'Domain', 'k': 1, 'positive label': 'Actinobacteria', 'label tax group': 'class', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'oldout'}\n",
      "24\n",
      "Actinobacteria class DNA_features 3 /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "start of making  training X,539, (539, 64)\n",
      "539 (539, 64)\n",
      "start of making  test X,210, (210, 64)\n",
      "210 (210, 64)\n",
      "labels y 0,539\n",
      "labels y 1,210\n",
      "X_train,Y_train (539, 64) (539,)\n",
      "length holdout data 138 138 138\n",
      "{'AUC': 0.999, 'accuracy': 0.99, 'specificity': 1.0, 'sensitivity': 0.98, 'TN/FP/FN/TP': array([110,   0,   2,  98]), 'N in class': 539, 'Features': 'DNA', 'k': 3, 'positive label': 'Actinobacteria', 'label tax group': 'class', 'pool label': 'Bacteria', 'pool tax group': 'kingdom', 'Baltimore': 'DNA', 'virus holdout group': 'Siphoviridae', 'training group': 'All'}\n",
      "Actinobacteria class DNA_features 6 /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "start of making  training X,539, (539, 4096)\n"
     ]
    }
   ],
   "source": [
    "for subset in subsets:\n",
    "#     # for holdout data  \n",
    "    datasets_ho,n1,n2 = get_class_lists_ho (subset,viruses,hosts)\n",
    "    test_dataset((*subset,'Holdout'),datasets_ho,n1)\n",
    "    \n",
    "    #for all data\n",
    "   \n",
    "    dataset2,ntrain,n_ho = get_class_lists(subset,viruses,hosts)\n",
    "    test_dataset((*subset,'All'), dataset2,ntrain)\n",
    "       "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get numbers for each data set in subsets. \n",
    "If too small remove"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check datasets reads OK and are sensilble\n",
    "for subset in subsets:\n",
    "    \n",
    "    class_data_ho,  ntest, n2 = get_class_lists_ho (subset,viruses,hosts)\n",
    "    class_data,n,nho = get_class_lists(subset,viruses,hosts)\n",
    "   # if nho<15 or n2 < 20:\n",
    "    print ('****too small',subset,n2,nho)\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Change case in subsetfiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "subsetfile = 'holdout_V_Bact2.csv'\n",
    "with open(subsetfile, 'w') as csvfile:\n",
    "    writer = csv.writer(csvfile)\n",
    "    writer.writerow(['label','label_tax','pool','pool_tax','baltimore','virus holdout'])\n",
    "    for subset in subsets:\n",
    "        sublist = list(subset)\n",
    "        sublist[1]= sublist[1].lower()\n",
    "        sublist[3]= sublist[3].lower()\n",
    "\n",
    "        writer.writerow(tuple (sublist)) "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
