{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "import os\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import random\n",
    "import numpy as np\n",
    "import timeit\n",
    "from collections import defaultdict, Counter\n",
    "from pathlib import Path\n",
    "from Bio import SeqIO\n",
    "import sys\n",
    "lib_dir = '../mylibs'\n",
    "if lib_dir not in sys.path:\n",
    "    sys.path.append(lib_dir)\n",
    "    \n",
    "import vhdb as vhdb\n",
    "import features_predict as fp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "filein = '../inputs/VHDB_25_1_2019.p'\n",
    "with open(filein, 'rb') as f:\n",
    "    V_H = pickle.load( f)\n",
    "hosts = V_H.hosts\n",
    "viruses = V_H.viruses\n",
    "subspecies =V_H.subspecies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results in Euk_all_results.csv\n"
     ]
    }
   ],
   "source": [
    "# Input file\n",
    "#subsets = [(label,taxlevel,pool,taxlevelpool),...]\n",
    "subsetfile = '../inputs/Euk_all.csv'\n",
    "baltimore = 'all'\n",
    "vhdbfile = '../inputs/VHDB_25_1_2019.p'\n",
    "\n",
    "\n",
    "# Everthing in the same order as the features list\n",
    "\n",
    "features = ['DNA','AA','PC','Domains']\n",
    "# filepath for the fasta, faa and domain files\n",
    "filepaths = [\n",
    "                '/home4/youn01f/Desktop/workspace/newData/fasta',\n",
    "                '/home4/youn01f/Desktop/workspace/newData/faa','/home4/youn01f/Desktop/workspace/newData/faa',\n",
    "                '/home4/youn01f/Desktop/workspace/newData/pfs']\n",
    "file_exts = ['fasta','fasta','fasta','pfs']#,'fasta',fasta]\n",
    "\n",
    "# A list of kmers for each feature set to be tested\n",
    "kmer_lists = [[1, 2,3,4,5,6,7,8,9], # dna \n",
    "              [1, 2, 3,4], # aa\n",
    "              [1, 2,3,4,5,6] ,    #pc\n",
    "             [0]] #domains\n",
    "\n",
    "# symbol dictionaries \n",
    "na_dict = {'mod':4,'a':0,'c':1,'g':2,'t':3}\n",
    "aa_dict = {'mod':20 ,'A':0,'C':1,'D':2,'E':3,'F':4,'G':5,'H':6,'I':7,'K':8,'L':9,'M':10,'N':11,\n",
    "              'P':12,'Q':13,'R':14,'S':15,'T':16,'V':17,'W':18,'Y':19}\n",
    "pc_dict = {'mod':7, 'C':0,\n",
    "            'A':1,'G':1,'V':1,\n",
    "           'I':2,'L':2,'F':2,'P':2,\n",
    "           'M':3,'S':3,'T':3,'Y':3,\n",
    "           'H':4,'N':4,'Q':4,'W':4,\n",
    "           'R':5,'K':5,\n",
    "           'D':6,'E':6}\n",
    "symbol_dicts = [ na_dict,aa_dict,pc_dict, {}]\n",
    "\n",
    "classifiers = ['SVM_lin']\n",
    "\n",
    "# Output file for the results\n",
    "\n",
    "results_file = f'{Path(subsetfile).stem}_results.csv'\n",
    "print( f'Results in {results_file}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "69"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "label_info = pd.read_csv(subsetfile)\n",
    "subsets = label_info.apply(tuple, axis=1).tolist()\n",
    "len(subsets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Bos taurus', 'species', 'Bovidae', 'family', 'all')\n",
      " size of training 277 and test set   71\n",
      "('Bos', 'genus', 'Bovidae', 'family', 'all')\n",
      " size of training 291 and test set   73\n",
      "('Ovis aries', 'species', 'Bovidae', 'family', 'all')\n",
      " size of training 67 and test set   19\n",
      "('Sus scrofa', 'species', 'Mammalia', 'class', 'all')\n",
      " size of training 301 and test set   73\n",
      "('Suidae', 'family', 'Mammalia', 'class', 'all')\n",
      " size of training 307 and test set   73\n",
      "('Vespertilionidae', 'family', 'Chiroptera', 'order', 'all')\n",
      " size of training 110 and test set   28\n",
      "('Phyllostomidae', 'family', 'Chiroptera', 'order', 'all')\n",
      " size of training 60 and test set   18\n",
      "('Rhinolophus', 'genus', 'Chiroptera', 'order', 'all')\n",
      " size of training 52 and test set   16\n",
      "('Pteropus', 'genus', 'Pteropodidae', 'family', 'all')\n",
      " size of training 48 and test set   16\n",
      "('Pteropodidae', 'family', 'Chiroptera', 'order', 'all')\n",
      " size of training 128 and test set   32\n",
      "('Chiroptera', 'order', 'Mammalia', 'class', 'all')\n",
      " size of training 402 and test set   98\n",
      "('Equus caballus', 'species', 'Mammalia', 'class', 'all')\n",
      " size of training 70 and test set   20\n",
      "('Equus', 'genus', 'Mammalia', 'class', 'all')\n",
      " size of training 88 and test set   18\n",
      "('Equidae', 'family', 'Mammalia', 'class', 'all')\n",
      " size of training 90 and test set   18\n",
      "('Perissodactyla', 'order', 'Mammalia', 'class', 'all')\n",
      " size of training 90 and test set   18\n",
      "('Canis lupus', 'species', 'Carnivora', 'order', 'all')\n",
      " size of training 88 and test set   18\n",
      "('Canidae', 'family', 'Carnivora', 'order', 'all')\n",
      " size of training 95 and test set   19\n",
      "('Felis catus', 'species', 'Carnivora', 'order', 'all')\n",
      " size of training 91 and test set   19\n",
      "('Felis', 'genus', 'Carnivora', 'order', 'all')\n",
      " size of training 91 and test set   19\n",
      "('Felidae', 'family', 'Carnivora', 'order', 'all')\n",
      " size of training 107 and test set   25\n",
      "('Carnivora', 'order', 'Mammalia', 'class', 'all')\n",
      " size of training 339 and test set   79\n",
      "('Eulipotyphla', 'order', 'Mammalia', 'class', 'all')\n",
      " size of training 50 and test set   16\n",
      "('Pan troglodytes', 'species', 'Hominidae', 'family', 'all')\n",
      " size of training 93 and test set   19\n",
      "('Pan', 'genus', 'Hominidae', 'family', 'all')\n",
      " size of training 106 and test set   22\n",
      "('Homo sapiens', 'species', 'Hominidae', 'family', 'all')\n",
      " size of training 402 and test set   98\n",
      "('Homo', 'genus', 'Hominidae', 'family', 'all')\n",
      " size of training 402 and test set   98\n",
      "('Hominidae', 'family', 'Primates', 'order', 'all')\n",
      " size of training 402 and test set   98\n",
      "('Macaca mulatta', 'species', 'Macaca', 'genus', 'all')\n",
      " size of training 90 and test set   18\n",
      "('Macaca', 'genus', 'Cercopithecidae', 'family', 'all')\n",
      " size of training 152 and test set   34\n",
      "('Chlorocebus aethiops', 'species', 'Cercopithecidae', 'family', 'all')\n",
      " size of training 107 and test set   25\n",
      "('Chlorocebus', 'genus', 'Cercopithecidae', 'family', 'all')\n",
      " size of training 120 and test set   30\n",
      "('Cercopithecidae', 'family', 'Primates', 'order', 'all')\n",
      " size of training 309 and test set   75\n",
      "('Primates', 'order', 'Mammalia', 'class', 'all')\n",
      " size of training 402 and test set   98\n",
      "('Mus musculus', 'species', 'Muridae', 'family', 'all')\n",
      " size of training 226 and test set   58\n",
      "('Mus', 'genus', 'Muridae', 'family', 'all')\n",
      " size of training 242 and test set   62\n",
      "('Rattus norvegicus', 'species', 'Muridae', 'family', 'all')\n",
      " size of training 65 and test set   19\n",
      "('Rattus', 'genus', 'Muridae', 'family', 'all')\n",
      " size of training 82 and test set   18\n",
      "('Muridae', 'family', 'Rodentia', 'order', 'all')\n",
      " size of training 373 and test set   89\n",
      "('Cricetidae', 'family', 'Rodentia', 'order', 'all')\n",
      " size of training 157 and test set   35\n",
      "('Rodentia', 'order', 'Mammalia', 'class', 'all')\n",
      " size of training 402 and test set   98\n",
      "('Mammalia', 'class', 'Chordata', 'phylum', 'all')\n",
      " size of training 402 and test set   98\n",
      "('Passeriformes', 'order', 'Aves', 'class', 'all')\n",
      " size of training 90 and test set   18\n",
      "('Meleagris gallopavo', 'species', 'Phasianidae', 'family', 'all')\n",
      " size of training 42 and test set   14\n",
      "('Gallus gallus', 'species', 'Galliformes', 'order', 'all')\n",
      " size of training 124 and test set   32\n",
      "('Phasianidae', 'family', 'Aves', 'class', 'all')\n",
      " size of training 169 and test set   39\n",
      "('Galliformes', 'order', 'Aves', 'class', 'all')\n",
      " size of training 182 and test set   46\n",
      "('Anas', 'genus', 'Aves', 'class', 'all')\n",
      " size of training 60 and test set   18\n",
      "('Anatidae', 'family', 'Aves', 'class', 'all')\n",
      " size of training 108 and test set   26\n",
      "('Aves', 'class', 'Chordata', 'phylum', 'all')\n",
      " size of training 402 and test set   98\n",
      "('Chordata', 'phylum', 'Metazoa', 'kingdom', 'all')\n",
      " size of training 402 and test set   98\n",
      "('Ixodidae', 'family', 'Arthropoda', 'phylum', 'all')\n",
      " size of training 78 and test set   20\n",
      "('Ixodida', 'order', 'Arthropoda', 'phylum', 'all')\n",
      " size of training 107 and test set   25\n",
      "('Arachnida', 'class', 'Arthropoda', 'phylum', 'all')\n",
      " size of training 120 and test set   32\n",
      "('Hymenoptera', 'order', 'Insecta', 'class', 'all')\n",
      " size of training 86 and test set   18\n",
      "('Aedes', 'genus', 'Culicidae', 'family', 'all')\n",
      " size of training 88 and test set   18\n",
      "('Culex', 'genus', 'Culicidae', 'family', 'all')\n",
      " size of training 122 and test set   32\n",
      "('Culicidae', 'family', 'Diptera', 'order', 'all')\n",
      " size of training 307 and test set   73\n",
      "('Diptera', 'order', 'Insecta', 'class', 'all')\n",
      " size of training 402 and test set   98\n",
      "('Noctuidae', 'family', 'Lepidoptera', 'order', 'all')\n",
      " size of training 90 and test set   18\n",
      "('Lepidoptera', 'order', 'Insecta', 'class', 'all')\n",
      " size of training 252 and test set   64\n",
      "('Hemiptera', 'order', 'Insecta', 'class', 'all')\n",
      " size of training 106 and test set   24\n",
      "('Libellulidae', 'family', 'Insecta', 'class', 'all')\n",
      " size of training 44 and test set   16\n",
      "('Odonata', 'order', 'Insecta', 'class', 'all')\n",
      " size of training 78 and test set   20\n",
      "('Insecta', 'class', 'Arthropoda', 'phylum', 'all')\n",
      " size of training 402 and test set   98\n",
      "('Decapoda', 'order', 'Arthropoda', 'phylum', 'all')\n",
      " size of training 52 and test set   16\n",
      "('Arthropoda', 'phylum', 'Metazoa', 'kingdom', 'all')\n",
      " size of training 402 and test set   98\n",
      "('Veneroida', 'order', 'Metazoa', 'kingdom', 'all')\n",
      " size of training 43 and test set   15\n",
      "('Bivalvia', 'class', 'Metazoa', 'kingdom', 'all')\n",
      " size of training 60 and test set   18\n",
      "('Mollusca', 'phylum', 'Metazoa', 'kingdom', 'all')\n",
      " size of training 88 and test set   18\n"
     ]
    }
   ],
   "source": [
    "for subset in subsets:\n",
    "    print (subset)\n",
    "    class_data,n = fp.get_class_lists(subset,viruses,hosts)\n",
    "    datasets = fp.split_data(viruses,class_data,n)\n",
    "    (label,label_tax,pool,pool_tax,baltimore) = subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_kmers (sequences,k,symbol_dict):\n",
    "    if k == 0: # domainsdom_list = list({dom for s in sequence.values() for dom in s})\n",
    "        \n",
    "        X,seq_index,f_index = get_domains(sequences,symbol_dict)\n",
    "    else:  # everything else\n",
    "        vocab_len =  symbol_dict['mod'] **k\n",
    "        f_index = get_feature_names(symbol_dict,k)\n",
    "        seq_index = []\n",
    "        X = np.zeros((len(sequences) ,vocab_len))\n",
    "\n",
    "        for i,(accn, seq) in enumerate (sequences.items()):\n",
    "            X[i] = get_word_frequ (seq,k, vocab_len,symbol_dict)\n",
    "            seq_index.append(accn)\n",
    "# normalise for length of genome(s)\n",
    "    XN = np.divide(X,X.sum(axis = 1)[:,None])\n",
    "    print (np.shape(XN)) \n",
    "    return XN,seq_index,f_index\n",
    "\n",
    "def get_domains(sequence,dom_list):\n",
    "    \"\"\"input: sequences dictionary of training and test sequence dicts\"\"\"\n",
    "    seq_index = []\n",
    "    X = np.zeros((len(sequence) ,len(dom_list)))\n",
    "    for i,(accn, seq) in enumerate (sequence.items()):\n",
    "        for dom in seq:\n",
    "            j = dom_list.index(dom)\n",
    "            X[i,j] += 1\n",
    "        seq_index.append(accn)\n",
    "    return X,seq_index,dom_list\n",
    "\n",
    "def get_word_frequ(sequence,k,vocab_len,symbol_dict):\n",
    "    \"\"\"returns the word frequency array for all genomes for a virus \"\"\"\n",
    "    word_freqs = np.zeros(shape =(vocab_len),dtype=np.int32)\n",
    "    for i in range(len(sequence)-k+1):\n",
    "        word = sequence[i:i+k]\n",
    "        if check_word(word,symbol_dict):\n",
    "            kmer_index = patternToNumber (word,k, symbol_dict)\n",
    "            word_freqs [ kmer_index] += 1\n",
    "            # If this for DNA/RNA then add the k-mers complinent\n",
    "#             if symbol_dict['mod'] == 4:\n",
    "#                 word_freqs [patternToNumber(compliment(word),k,symbol_dict)] +=1   \n",
    "    return word_freqs\n",
    "\n",
    "def get_sequences(faapath,datasets,ext):\n",
    "    missing_files = []\n",
    "    v_with_missing_files = []\n",
    "    ext = f'.{ext}'\n",
    "    sequences = {'training':{},'test':{}}\n",
    "    for key,dataset in datasets.items():\n",
    "        for v,vdict in dataset.items():\n",
    "            sequence = defaultdict(str)\n",
    "            for ref in vdict['refseqs']:\n",
    "                filename = os.path.join(faapath, ref.strip()+ ext)\n",
    "                try:\n",
    "                    with open(filename, 'rt') as handle:\n",
    "                        if ext == '.fasta':\n",
    "                            for record in SeqIO.parse(handle, \"fasta\"):                   \n",
    "                                sequence[v] += str(record.seq)\n",
    "                        else: # pfs file - list of domains  \n",
    "                            line = handle.read()\n",
    "                            sequence[v] = [d for d in line.split()]\n",
    "                                \n",
    "                except IOError:\n",
    "                    missing_files.append((v,ref))\n",
    "                    \n",
    "            if len(sequence) > 0:\n",
    "                sequences[key].update(sequence)\n",
    "            else:\n",
    "                v_with_missing_files.append(v)\n",
    "    print ( f'viruses no files {len(set(v_with_missing_files))}')\n",
    "    return (sequences)\n",
    "\n",
    "\n",
    "def symbolToNumber (symbol,symbols):\n",
    "    return symbols[symbol]\n",
    "\n",
    "def patternToNumber (pattern,k,symbol_dict):\n",
    "    mod = symbol_dict['mod']\n",
    "    number = 0\n",
    "    for i in range (0,k):\n",
    "        n = symbolToNumber ( pattern [k-1-i],symbol_dict)\n",
    "        number += n* mod**i \n",
    "    return number\n",
    "\n",
    "def get_feature_names(symbols,k):\n",
    "    all_kmers = []\n",
    "    n2sym = { v:k for k,v in symbols.items()}\n",
    "    mod = symbols['mod']\n",
    "    for number in range(mod**k):\n",
    "        pattern = ''\n",
    "        for i in range (0,k):\n",
    "            n = number% mod\n",
    "            number = (number - n)/mod \n",
    "            symbol = n2sym[n]\n",
    "            pattern = symbol + pattern\n",
    "        all_kmers.append (pattern)\n",
    "    return all_kmers\n",
    "\n",
    "def check_word(mystring,symbols):\n",
    "    \"\"\"Check for illegal characters \"\"\"\n",
    "    return all(c in symbols for c in mystring)\n",
    "\n",
    "def compliment(dna):\n",
    "    \"\"\"Return the reverse compliment of a sequence \"\"\"\n",
    "    bp = { 'a':'t', 't':'a', 'c':'g','g':'c'}\n",
    "    revdna =''    \n",
    "    for i in range(len(dna)-1,-1,-1):\n",
    "         revdna += bp[dna[i]]\n",
    "    return (revdna)\n",
    "\n",
    "def  get_labels(label,v_index,dataset):      \n",
    "    for i,v in enumerate(v_index):\n",
    "        y = [1 if dataset[v]['label'] ==  label else 0 for v in v_index ] \n",
    "    return np.asarray (y)\n",
    "\n",
    "def get_feature_matrices(sequences,datasets,k,symbol_dict):\n",
    "   \n",
    "    if k == 0: # domainsdom_list = list({dom for s in sequence.values() for dom in s})\n",
    "        symbol_dict = list({dom for sequence in sequences.values() for s in sequence.values() for dom in s}) \n",
    "    print (f'X_training:Extacting  {feature}  of length  {k}  from   {len(sequences[\"training\"])} sequences')\n",
    "    X_train, seq_index, f_index = extract_kmers (sequences['training'],k,symbol_dict)\n",
    "    Y_train =  get_labels (label,seq_index,datasets['training'])\n",
    "    print (f'X_test: Extacting  {feature}  of length  {k}  from   {len(sequences[\"test\"])} sequences')\n",
    "    X_test, seq_index, f_index = extract_kmers (sequences['test'],k,symbol_dict)\n",
    "    Y_test =  get_labels (label,seq_index,datasets['test'])\n",
    "           \n",
    "    return X_train,X_test,Y_train,Y_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main Loop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " size of training 106 and test set   24\n",
      "Bos taurus species Bovidae family all\n",
      "0 [1, 2, 3, 4, 5, 6, 7, 8, 9] /home4/youn01f/Desktop/workspace/newData/fasta {'mod': 4, 'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
      "getting DNA sequences from /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "viruses no files 4\n",
      "X_training:Extacting  DNA  of length  1  from   103 sequences\n",
      "(103, 4)\n",
      "X_test: Extacting  DNA  of length  1  from   23 sequences\n",
      "(23, 4)\n",
      "(103, 4) (23, 4) (103,) (23,)\n",
      "{'AUC': 0.341, 'N in class': 65, 'Features': 'DNA', 'k': 1}\n",
      "X_training:Extacting  DNA  of length  2  from   103 sequences\n",
      "(103, 16)\n",
      "X_test: Extacting  DNA  of length  2  from   23 sequences\n",
      "(23, 16)\n",
      "(103, 16) (23, 16) (103,) (23,)\n",
      "{'AUC': 0.286, 'N in class': 65, 'Features': 'DNA', 'k': 2}\n",
      "X_training:Extacting  DNA  of length  3  from   103 sequences\n",
      "(103, 64)\n",
      "X_test: Extacting  DNA  of length  3  from   23 sequences\n",
      "(23, 64)\n",
      "(103, 64) (23, 64) (103,) (23,)\n",
      "{'AUC': 0.587, 'N in class': 65, 'Features': 'DNA', 'k': 3}\n",
      "X_training:Extacting  DNA  of length  4  from   103 sequences\n",
      "(103, 256)\n",
      "X_test: Extacting  DNA  of length  4  from   23 sequences\n",
      "(23, 256)\n",
      "(103, 256) (23, 256) (103,) (23,)\n",
      "{'AUC': 0.73, 'N in class': 65, 'Features': 'DNA', 'k': 4}\n",
      "X_training:Extacting  DNA  of length  5  from   103 sequences\n",
      "(103, 1024)\n",
      "X_test: Extacting  DNA  of length  5  from   23 sequences\n",
      "(23, 1024)\n",
      "(103, 1024) (23, 1024) (103,) (23,)\n",
      "{'AUC': 0.738, 'N in class': 65, 'Features': 'DNA', 'k': 5}\n",
      "X_training:Extacting  DNA  of length  6  from   103 sequences\n",
      "(103, 4096)\n",
      "X_test: Extacting  DNA  of length  6  from   23 sequences\n",
      "(23, 4096)\n",
      "(103, 4096) (23, 4096) (103,) (23,)\n",
      "{'AUC': 0.69, 'N in class': 65, 'Features': 'DNA', 'k': 6}\n",
      "X_training:Extacting  DNA  of length  7  from   103 sequences\n",
      "(103, 16384)\n",
      "X_test: Extacting  DNA  of length  7  from   23 sequences\n",
      "(23, 16384)\n",
      "(103, 16384) (23, 16384) (103,) (23,)\n",
      "{'AUC': 0.722, 'N in class': 65, 'Features': 'DNA', 'k': 7}\n",
      "X_training:Extacting  DNA  of length  8  from   103 sequences\n",
      "(103, 65536)\n",
      "X_test: Extacting  DNA  of length  8  from   23 sequences\n",
      "(23, 65536)\n",
      "(103, 65536) (23, 65536) (103,) (23,)\n",
      "{'AUC': 0.762, 'N in class': 65, 'Features': 'DNA', 'k': 8}\n",
      "X_training:Extacting  DNA  of length  9  from   103 sequences\n",
      "(103, 262144)\n",
      "X_test: Extacting  DNA  of length  9  from   23 sequences\n",
      "(23, 262144)\n",
      "(103, 262144) (23, 262144) (103,) (23,)\n",
      "{'AUC': 0.706, 'N in class': 65, 'Features': 'DNA', 'k': 9}\n",
      "1 [1, 2, 3, 4] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 20, 'A': 0, 'C': 1, 'D': 2, 'E': 3, 'F': 4, 'G': 5, 'H': 6, 'I': 7, 'K': 8, 'L': 9, 'M': 10, 'N': 11, 'P': 12, 'Q': 13, 'R': 14, 'S': 15, 'T': 16, 'V': 17, 'W': 18, 'Y': 19}\n",
      "getting AA sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "viruses no files 0\n",
      "X_training:Extacting  AA  of length  1  from   106 sequences\n",
      "(106, 20)\n",
      "X_test: Extacting  AA  of length  1  from   24 sequences\n",
      "(24, 20)\n",
      "(106, 20) (24, 20) (106,) (24,)\n",
      "{'AUC': 0.607, 'N in class': 65, 'Features': 'AA', 'k': 1}\n",
      "X_training:Extacting  AA  of length  2  from   106 sequences\n",
      "(106, 400)\n",
      "X_test: Extacting  AA  of length  2  from   24 sequences\n",
      "(24, 400)\n",
      "(106, 400) (24, 400) (106,) (24,)\n",
      "{'AUC': 0.539, 'N in class': 65, 'Features': 'AA', 'k': 2}\n",
      "X_training:Extacting  AA  of length  3  from   106 sequences\n",
      "(106, 8000)\n",
      "X_test: Extacting  AA  of length  3  from   24 sequences\n",
      "(24, 8000)\n",
      "(106, 8000) (24, 8000) (106,) (24,)\n",
      "{'AUC': 0.729, 'N in class': 65, 'Features': 'AA', 'k': 3}\n",
      "X_training:Extacting  AA  of length  4  from   106 sequences\n",
      "(106, 160000)\n",
      "X_test: Extacting  AA  of length  4  from   24 sequences\n",
      "(24, 160000)\n",
      "(106, 160000) (24, 160000) (106,) (24,)\n",
      "{'AUC': 0.85, 'N in class': 65, 'Features': 'AA', 'k': 4}\n",
      "2 [1, 2, 3, 4, 5, 6] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 7, 'C': 0, 'A': 1, 'G': 1, 'V': 1, 'I': 2, 'L': 2, 'F': 2, 'P': 2, 'M': 3, 'S': 3, 'T': 3, 'Y': 3, 'H': 4, 'N': 4, 'Q': 4, 'W': 4, 'R': 5, 'K': 5, 'D': 6, 'E': 6}\n",
      "getting PC sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "viruses no files 0\n",
      "X_training:Extacting  PC  of length  1  from   106 sequences\n",
      "(106, 7)\n",
      "X_test: Extacting  PC  of length  1  from   24 sequences\n",
      "(24, 7)\n",
      "(106, 7) (24, 7) (106,) (24,)\n",
      "{'AUC': 0.4, 'N in class': 65, 'Features': 'PC', 'k': 1}\n",
      "X_training:Extacting  PC  of length  2  from   106 sequences\n",
      "(106, 49)\n",
      "X_test: Extacting  PC  of length  2  from   24 sequences\n",
      "(24, 49)\n",
      "(106, 49) (24, 49) (106,) (24,)\n",
      "{'AUC': 0.743, 'N in class': 65, 'Features': 'PC', 'k': 2}\n",
      "X_training:Extacting  PC  of length  3  from   106 sequences\n",
      "(106, 343)\n",
      "X_test: Extacting  PC  of length  3  from   24 sequences\n",
      "(24, 343)\n",
      "(106, 343) (24, 343) (106,) (24,)\n",
      "{'AUC': 0.539, 'N in class': 65, 'Features': 'PC', 'k': 3}\n",
      "X_training:Extacting  PC  of length  4  from   106 sequences\n",
      "(106, 2401)\n",
      "X_test: Extacting  PC  of length  4  from   24 sequences\n",
      "(24, 2401)\n",
      "(106, 2401) (24, 2401) (106,) (24,)\n",
      "{'AUC': 0.421, 'N in class': 65, 'Features': 'PC', 'k': 4}\n",
      "X_training:Extacting  PC  of length  5  from   106 sequences\n",
      "(106, 16807)\n",
      "X_test: Extacting  PC  of length  5  from   24 sequences\n",
      "(24, 16807)\n",
      "(106, 16807) (24, 16807) (106,) (24,)\n",
      "{'AUC': 0.771, 'N in class': 65, 'Features': 'PC', 'k': 5}\n",
      "X_training:Extacting  PC  of length  6  from   106 sequences\n",
      "(106, 117649)\n",
      "X_test: Extacting  PC  of length  6  from   24 sequences\n",
      "(24, 117649)\n",
      "(106, 117649) (24, 117649) (106,) (24,)\n",
      "{'AUC': 0.85, 'N in class': 65, 'Features': 'PC', 'k': 6}\n",
      "3 [0] /home4/youn01f/Desktop/workspace/newData/pfs {}\n",
      "getting Domains sequences from /home4/youn01f/Desktop/workspace/newData/pfs\n",
      "viruses no files 8\n",
      "X_training:Extacting  Domains  of length  0  from   98 sequences\n",
      "(98, 431)\n",
      "X_test: Extacting  Domains  of length  0  from   24 sequences\n",
      "(24, 431)\n",
      "(98, 431) (24, 431) (98,) (24,)\n",
      "{'AUC': 0.793, 'N in class': 65, 'Features': 'Domains', 'k': 0}\n",
      " size of training 95 and test set   19\n",
      "Bos genus Bovidae family all\n",
      "0 [1, 2, 3, 4, 5, 6, 7, 8, 9] /home4/youn01f/Desktop/workspace/newData/fasta {'mod': 4, 'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
      "getting DNA sequences from /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "viruses no files 2\n",
      "X_training:Extacting  DNA  of length  1  from   93 sequences\n",
      "(93, 4)\n",
      "X_test: Extacting  DNA  of length  1  from   19 sequences\n",
      "(19, 4)\n",
      "(93, 4) (19, 4) (93,) (19,)\n",
      "{'AUC': 0.375, 'N in class': 57, 'Features': 'DNA', 'k': 1}\n",
      "X_training:Extacting  DNA  of length  2  from   93 sequences\n",
      "(93, 16)\n",
      "X_test: Extacting  DNA  of length  2  from   19 sequences\n",
      "(19, 16)\n",
      "(93, 16) (19, 16) (93,) (19,)\n",
      "{'AUC': 0.452, 'N in class': 57, 'Features': 'DNA', 'k': 2}\n",
      "X_training:Extacting  DNA  of length  3  from   93 sequences\n",
      "(93, 64)\n",
      "X_test: Extacting  DNA  of length  3  from   19 sequences\n",
      "(19, 64)\n",
      "(93, 64) (19, 64) (93,) (19,)\n",
      "{'AUC': 0.393, 'N in class': 57, 'Features': 'DNA', 'k': 3}\n",
      "X_training:Extacting  DNA  of length  4  from   93 sequences\n",
      "(93, 256)\n",
      "X_test: Extacting  DNA  of length  4  from   19 sequences\n",
      "(19, 256)\n",
      "(93, 256) (19, 256) (93,) (19,)\n",
      "{'AUC': 0.619, 'N in class': 57, 'Features': 'DNA', 'k': 4}\n",
      "X_training:Extacting  DNA  of length  5  from   93 sequences\n",
      "(93, 1024)\n",
      "X_test: Extacting  DNA  of length  5  from   19 sequences\n",
      "(19, 1024)\n",
      "(93, 1024) (19, 1024) (93,) (19,)\n",
      "{'AUC': 0.833, 'N in class': 57, 'Features': 'DNA', 'k': 5}\n",
      "X_training:Extacting  DNA  of length  6  from   93 sequences\n",
      "(93, 4096)\n",
      "X_test: Extacting  DNA  of length  6  from   19 sequences\n",
      "(19, 4096)\n",
      "(93, 4096) (19, 4096) (93,) (19,)\n",
      "{'AUC': 0.702, 'N in class': 57, 'Features': 'DNA', 'k': 6}\n",
      "X_training:Extacting  DNA  of length  7  from   93 sequences\n",
      "(93, 16384)\n",
      "X_test: Extacting  DNA  of length  7  from   19 sequences\n",
      "(19, 16384)\n",
      "(93, 16384) (19, 16384) (93,) (19,)\n",
      "{'AUC': 0.327, 'N in class': 57, 'Features': 'DNA', 'k': 7}\n",
      "X_training:Extacting  DNA  of length  8  from   93 sequences\n",
      "(93, 65536)\n",
      "X_test: Extacting  DNA  of length  8  from   19 sequences\n",
      "(19, 65536)\n",
      "(93, 65536) (19, 65536) (93,) (19,)\n",
      "{'AUC': 0.19, 'N in class': 57, 'Features': 'DNA', 'k': 8}\n",
      "X_training:Extacting  DNA  of length  9  from   93 sequences\n",
      "(93, 262144)\n",
      "X_test: Extacting  DNA  of length  9  from   19 sequences\n",
      "(19, 262144)\n",
      "(93, 262144) (19, 262144) (93,) (19,)\n",
      "{'AUC': 0.929, 'N in class': 57, 'Features': 'DNA', 'k': 9}\n",
      "1 [1, 2, 3, 4] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 20, 'A': 0, 'C': 1, 'D': 2, 'E': 3, 'F': 4, 'G': 5, 'H': 6, 'I': 7, 'K': 8, 'L': 9, 'M': 10, 'N': 11, 'P': 12, 'Q': 13, 'R': 14, 'S': 15, 'T': 16, 'V': 17, 'W': 18, 'Y': 19}\n",
      "getting AA sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "viruses no files 0\n",
      "X_training:Extacting  AA  of length  1  from   95 sequences\n",
      "(95, 20)\n",
      "X_test: Extacting  AA  of length  1  from   19 sequences\n",
      "(19, 20)\n",
      "(95, 20) (19, 20) (95,) (19,)\n",
      "{'AUC': 0.512, 'N in class': 57, 'Features': 'AA', 'k': 1}\n",
      "X_training:Extacting  AA  of length  2  from   95 sequences\n",
      "(95, 400)\n",
      "X_test: Extacting  AA  of length  2  from   19 sequences\n",
      "(19, 400)\n",
      "(95, 400) (19, 400) (95,) (19,)\n",
      "{'AUC': 0.821, 'N in class': 57, 'Features': 'AA', 'k': 2}\n",
      "X_training:Extacting  AA  of length  3  from   95 sequences\n",
      "(95, 8000)\n",
      "X_test: Extacting  AA  of length  3  from   19 sequences\n",
      "(19, 8000)\n",
      "(95, 8000) (19, 8000) (95,) (19,)\n",
      "{'AUC': 0.798, 'N in class': 57, 'Features': 'AA', 'k': 3}\n",
      "X_training:Extacting  AA  of length  4  from   95 sequences\n",
      "(95, 160000)\n",
      "X_test: Extacting  AA  of length  4  from   19 sequences\n",
      "(19, 160000)\n",
      "(95, 160000) (19, 160000) (95,) (19,)\n",
      "{'AUC': 0.845, 'N in class': 57, 'Features': 'AA', 'k': 4}\n",
      "2 [1, 2, 3, 4, 5, 6] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 7, 'C': 0, 'A': 1, 'G': 1, 'V': 1, 'I': 2, 'L': 2, 'F': 2, 'P': 2, 'M': 3, 'S': 3, 'T': 3, 'Y': 3, 'H': 4, 'N': 4, 'Q': 4, 'W': 4, 'R': 5, 'K': 5, 'D': 6, 'E': 6}\n",
      "getting PC sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "viruses no files 0\n",
      "X_training:Extacting  PC  of length  1  from   95 sequences\n",
      "(95, 7)\n",
      "X_test: Extacting  PC  of length  1  from   19 sequences\n",
      "(19, 7)\n",
      "(95, 7) (19, 7) (95,) (19,)\n",
      "{'AUC': 0.333, 'N in class': 57, 'Features': 'PC', 'k': 1}\n",
      "X_training:Extacting  PC  of length  2  from   95 sequences\n",
      "(95, 49)\n",
      "X_test: Extacting  PC  of length  2  from   19 sequences\n",
      "(19, 49)\n",
      "(95, 49) (19, 49) (95,) (19,)\n",
      "{'AUC': 0.75, 'N in class': 57, 'Features': 'PC', 'k': 2}\n",
      "X_training:Extacting  PC  of length  3  from   95 sequences\n",
      "(95, 343)\n",
      "X_test: Extacting  PC  of length  3  from   19 sequences\n",
      "(19, 343)\n",
      "(95, 343) (19, 343) (95,) (19,)\n",
      "{'AUC': 0.702, 'N in class': 57, 'Features': 'PC', 'k': 3}\n",
      "X_training:Extacting  PC  of length  4  from   95 sequences\n",
      "(95, 2401)\n",
      "X_test: Extacting  PC  of length  4  from   19 sequences\n",
      "(19, 2401)\n",
      "(95, 2401) (19, 2401) (95,) (19,)\n",
      "{'AUC': 0.673, 'N in class': 57, 'Features': 'PC', 'k': 4}\n",
      "X_training:Extacting  PC  of length  5  from   95 sequences\n",
      "(95, 16807)\n",
      "X_test: Extacting  PC  of length  5  from   19 sequences\n",
      "(19, 16807)\n",
      "(95, 16807) (19, 16807) (95,) (19,)\n",
      "{'AUC': 0.667, 'N in class': 57, 'Features': 'PC', 'k': 5}\n",
      "X_training:Extacting  PC  of length  6  from   95 sequences\n",
      "(95, 117649)\n",
      "X_test: Extacting  PC  of length  6  from   19 sequences\n",
      "(19, 117649)\n",
      "(95, 117649) (19, 117649) (95,) (19,)\n",
      "{'AUC': 0.881, 'N in class': 57, 'Features': 'PC', 'k': 6}\n",
      "3 [0] /home4/youn01f/Desktop/workspace/newData/pfs {}\n",
      "getting Domains sequences from /home4/youn01f/Desktop/workspace/newData/pfs\n",
      "viruses no files 5\n",
      "X_training:Extacting  Domains  of length  0  from   93 sequences\n",
      "(93, 361)\n",
      "X_test: Extacting  Domains  of length  0  from   16 sequences\n",
      "(16, 361)\n",
      "(93, 361) (16, 361) (93,) (16,)\n",
      "{'AUC': 0.767, 'N in class': 57, 'Features': 'Domains', 'k': 0}\n",
      " size of training 67 and test set   19\n",
      "Ovis aries species Bovidae family all\n",
      "0 [1, 2, 3, 4, 5, 6, 7, 8, 9] /home4/youn01f/Desktop/workspace/newData/fasta {'mod': 4, 'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
      "getting DNA sequences from /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "viruses no files 2\n",
      "X_training:Extacting  DNA  of length  1  from   65 sequences\n",
      "(65, 4)\n",
      "X_test: Extacting  DNA  of length  1  from   19 sequences\n",
      "(19, 4)\n",
      "(65, 4) (19, 4) (65,) (19,)\n",
      "{'AUC': 0.682, 'N in class': 43, 'Features': 'DNA', 'k': 1}\n",
      "X_training:Extacting  DNA  of length  2  from   65 sequences\n",
      "(65, 16)\n",
      "X_test: Extacting  DNA  of length  2  from   19 sequences\n",
      "(19, 16)\n",
      "(65, 16) (19, 16) (65,) (19,)\n",
      "{'AUC': 0.682, 'N in class': 43, 'Features': 'DNA', 'k': 2}\n",
      "X_training:Extacting  DNA  of length  3  from   65 sequences\n",
      "(65, 64)\n",
      "X_test: Extacting  DNA  of length  3  from   19 sequences\n",
      "(19, 64)\n",
      "(65, 64) (19, 64) (65,) (19,)\n",
      "{'AUC': 0.239, 'N in class': 43, 'Features': 'DNA', 'k': 3}\n",
      "X_training:Extacting  DNA  of length  4  from   65 sequences\n",
      "(65, 256)\n",
      "X_test: Extacting  DNA  of length  4  from   19 sequences\n",
      "(19, 256)\n",
      "(65, 256) (19, 256) (65,) (19,)\n",
      "{'AUC': 0.409, 'N in class': 43, 'Features': 'DNA', 'k': 4}\n",
      "X_training:Extacting  DNA  of length  5  from   65 sequences\n",
      "(65, 1024)\n",
      "X_test: Extacting  DNA  of length  5  from   19 sequences\n",
      "(19, 1024)\n",
      "(65, 1024) (19, 1024) (65,) (19,)\n",
      "{'AUC': 0.398, 'N in class': 43, 'Features': 'DNA', 'k': 5}\n",
      "X_training:Extacting  DNA  of length  6  from   65 sequences\n",
      "(65, 4096)\n",
      "X_test: Extacting  DNA  of length  6  from   19 sequences\n",
      "(19, 4096)\n",
      "(65, 4096) (19, 4096) (65,) (19,)\n",
      "{'AUC': 0.352, 'N in class': 43, 'Features': 'DNA', 'k': 6}\n",
      "X_training:Extacting  DNA  of length  7  from   65 sequences\n",
      "(65, 16384)\n",
      "X_test: Extacting  DNA  of length  7  from   19 sequences\n",
      "(19, 16384)\n",
      "(65, 16384) (19, 16384) (65,) (19,)\n",
      "{'AUC': 0.398, 'N in class': 43, 'Features': 'DNA', 'k': 7}\n",
      "X_training:Extacting  DNA  of length  8  from   65 sequences\n",
      "(65, 65536)\n",
      "X_test: Extacting  DNA  of length  8  from   19 sequences\n",
      "(19, 65536)\n",
      "(65, 65536) (19, 65536) (65,) (19,)\n",
      "{'AUC': 0.682, 'N in class': 43, 'Features': 'DNA', 'k': 8}\n",
      "X_training:Extacting  DNA  of length  9  from   65 sequences\n",
      "(65, 262144)\n",
      "X_test: Extacting  DNA  of length  9  from   19 sequences\n",
      "(19, 262144)\n",
      "(65, 262144) (19, 262144) (65,) (19,)\n",
      "{'AUC': 0.727, 'N in class': 43, 'Features': 'DNA', 'k': 9}\n",
      "1 [1, 2, 3, 4] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 20, 'A': 0, 'C': 1, 'D': 2, 'E': 3, 'F': 4, 'G': 5, 'H': 6, 'I': 7, 'K': 8, 'L': 9, 'M': 10, 'N': 11, 'P': 12, 'Q': 13, 'R': 14, 'S': 15, 'T': 16, 'V': 17, 'W': 18, 'Y': 19}\n",
      "getting AA sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "viruses no files 0\n",
      "X_training:Extacting  AA  of length  1  from   67 sequences\n",
      "(67, 20)\n",
      "X_test: Extacting  AA  of length  1  from   19 sequences\n",
      "(19, 20)\n",
      "(67, 20) (19, 20) (67,) (19,)\n",
      "{'AUC': 0.42, 'N in class': 43, 'Features': 'AA', 'k': 1}\n",
      "X_training:Extacting  AA  of length  2  from   67 sequences\n",
      "(67, 400)\n",
      "X_test: Extacting  AA  of length  2  from   19 sequences\n",
      "(19, 400)\n",
      "(67, 400) (19, 400) (67,) (19,)\n",
      "{'AUC': 0.761, 'N in class': 43, 'Features': 'AA', 'k': 2}\n",
      "X_training:Extacting  AA  of length  3  from   67 sequences\n",
      "(67, 8000)\n",
      "X_test: Extacting  AA  of length  3  from   19 sequences\n",
      "(19, 8000)\n",
      "(67, 8000) (19, 8000) (67,) (19,)\n",
      "{'AUC': 0.489, 'N in class': 43, 'Features': 'AA', 'k': 3}\n",
      "X_training:Extacting  AA  of length  4  from   67 sequences\n",
      "(67, 160000)\n",
      "X_test: Extacting  AA  of length  4  from   19 sequences\n",
      "(19, 160000)\n",
      "(67, 160000) (19, 160000) (67,) (19,)\n",
      "{'AUC': 0.773, 'N in class': 43, 'Features': 'AA', 'k': 4}\n",
      "2 [1, 2, 3, 4, 5, 6] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 7, 'C': 0, 'A': 1, 'G': 1, 'V': 1, 'I': 2, 'L': 2, 'F': 2, 'P': 2, 'M': 3, 'S': 3, 'T': 3, 'Y': 3, 'H': 4, 'N': 4, 'Q': 4, 'W': 4, 'R': 5, 'K': 5, 'D': 6, 'E': 6}\n",
      "getting PC sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "viruses no files 0\n",
      "X_training:Extacting  PC  of length  1  from   67 sequences\n",
      "(67, 7)\n",
      "X_test: Extacting  PC  of length  1  from   19 sequences\n",
      "(19, 7)\n",
      "(67, 7) (19, 7) (67,) (19,)\n",
      "{'AUC': 0.33, 'N in class': 43, 'Features': 'PC', 'k': 1}\n",
      "X_training:Extacting  PC  of length  2  from   67 sequences\n",
      "(67, 49)\n",
      "X_test: Extacting  PC  of length  2  from   19 sequences\n",
      "(19, 49)\n",
      "(67, 49) (19, 49) (67,) (19,)\n",
      "{'AUC': 0.739, 'N in class': 43, 'Features': 'PC', 'k': 2}\n",
      "X_training:Extacting  PC  of length  3  from   67 sequences\n",
      "(67, 343)\n",
      "X_test: Extacting  PC  of length  3  from   19 sequences\n",
      "(19, 343)\n",
      "(67, 343) (19, 343) (67,) (19,)\n",
      "{'AUC': 0.602, 'N in class': 43, 'Features': 'PC', 'k': 3}\n",
      "X_training:Extacting  PC  of length  4  from   67 sequences\n",
      "(67, 2401)\n",
      "X_test: Extacting  PC  of length  4  from   19 sequences\n",
      "(19, 2401)\n",
      "(67, 2401) (19, 2401) (67,) (19,)\n",
      "{'AUC': 0.42, 'N in class': 43, 'Features': 'PC', 'k': 4}\n",
      "X_training:Extacting  PC  of length  5  from   67 sequences\n",
      "(67, 16807)\n",
      "X_test: Extacting  PC  of length  5  from   19 sequences\n",
      "(19, 16807)\n",
      "(67, 16807) (19, 16807) (67,) (19,)\n",
      "{'AUC': 0.688, 'N in class': 43, 'Features': 'PC', 'k': 5}\n",
      "X_training:Extacting  PC  of length  6  from   67 sequences\n",
      "(67, 117649)\n",
      "X_test: Extacting  PC  of length  6  from   19 sequences\n",
      "(19, 117649)\n",
      "(67, 117649) (19, 117649) (67,) (19,)\n",
      "{'AUC': 0.67, 'N in class': 43, 'Features': 'PC', 'k': 6}\n",
      "3 [0] /home4/youn01f/Desktop/workspace/newData/pfs {}\n",
      "getting Domains sequences from /home4/youn01f/Desktop/workspace/newData/pfs\n",
      "viruses no files 4\n",
      "X_training:Extacting  Domains  of length  0  from   64 sequences\n",
      "(64, 394)\n",
      "X_test: Extacting  Domains  of length  0  from   18 sequences\n",
      "(18, 394)\n",
      "(64, 394) (18, 394) (64,) (18,)\n",
      "{'AUC': 0.375, 'N in class': 43, 'Features': 'Domains', 'k': 0}\n",
      " size of training 301 and test set   73\n",
      "Sus scrofa species Mammalia class all\n",
      "0 [1, 2, 3, 4, 5, 6, 7, 8, 9] /home4/youn01f/Desktop/workspace/newData/fasta {'mod': 4, 'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
      "getting DNA sequences from /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "viruses no files 7\n",
      "X_training:Extacting  DNA  of length  1  from   295 sequences\n",
      "(295, 4)\n",
      "X_test: Extacting  DNA  of length  1  from   72 sequences\n",
      "(72, 4)\n",
      "(295, 4) (72, 4) (295,) (72,)\n",
      "{'AUC': 0.633, 'N in class': 187, 'Features': 'DNA', 'k': 1}\n",
      "X_training:Extacting  DNA  of length  2  from   295 sequences\n",
      "(295, 16)\n",
      "X_test: Extacting  DNA  of length  2  from   72 sequences\n",
      "(72, 16)\n",
      "(295, 16) (72, 16) (295,) (72,)\n",
      "{'AUC': 0.758, 'N in class': 187, 'Features': 'DNA', 'k': 2}\n",
      "X_training:Extacting  DNA  of length  3  from   295 sequences\n",
      "(295, 64)\n",
      "X_test: Extacting  DNA  of length  3  from   72 sequences\n",
      "(72, 64)\n",
      "(295, 64) (72, 64) (295,) (72,)\n",
      "{'AUC': 0.813, 'N in class': 187, 'Features': 'DNA', 'k': 3}\n",
      "X_training:Extacting  DNA  of length  4  from   295 sequences\n",
      "(295, 256)\n",
      "X_test: Extacting  DNA  of length  4  from   72 sequences\n",
      "(72, 256)\n",
      "(295, 256) (72, 256) (295,) (72,)\n",
      "{'AUC': 0.807, 'N in class': 187, 'Features': 'DNA', 'k': 4}\n",
      "X_training:Extacting  DNA  of length  5  from   295 sequences\n",
      "(295, 1024)\n",
      "X_test: Extacting  DNA  of length  5  from   72 sequences\n",
      "(72, 1024)\n",
      "(295, 1024) (72, 1024) (295,) (72,)\n",
      "{'AUC': 0.813, 'N in class': 187, 'Features': 'DNA', 'k': 5}\n",
      "X_training:Extacting  DNA  of length  6  from   295 sequences\n"
     ]
    }
   ],
   "source": [
    "for subset in subsets:\n",
    "    class_data,n = fp.get_class_lists(subset,viruses,hosts)\n",
    "    datasets = fp.split_data(viruses,class_data,n)\n",
    "    (label,label_tax,pool,pool_tax,baltimore) = subset\n",
    "    print  (label,label_tax,pool,pool_tax,baltimore)\n",
    "   \n",
    "    for feature in features:\n",
    "        \n",
    "        index = features.index(feature)\n",
    "        kmers = kmer_lists[index]\n",
    "        filepath = filepaths[index]\n",
    "        symbol_dict = symbol_dicts[index]\n",
    "        ext = file_exts[index]\n",
    "        print (index,kmers,filepath,symbol_dict)\n",
    "\n",
    "        if features != 'PC': #PC same sequence as for AA\n",
    "            filepath = filepaths[index]\n",
    "            print ( f'getting {feature} sequences from {filepath}')\n",
    "            sequences = get_sequences(filepath,datasets,ext)\n",
    "\n",
    "        for k in kmers: \n",
    "            x_train,x_test,y_train,y_test = get_feature_matrices(sequences,datasets,k,symbol_dict)\n",
    "            print (np.shape(x_train), np.shape(x_test),np.shape(y_train),np.shape(y_test))\n",
    "            #print (type(x_train), type(x_test),type(y_train),type(y_test))\n",
    "            results = fp.test_prediction(x_train,x_test,y_train,y_test)\n",
    "            results.update ({'N in class': n, 'Features':feature, 'k':k})\n",
    "            print(results)\n",
    "            fp.results2CSV (results,subset, results_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test one subset on on feature type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(datasets['training'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_list = ['DNA','AA','PC','Domains']\n",
    "features = ['AA']\n",
    "for feature in features:\n",
    "    index = feature_list.index(feature)\n",
    "    kmers = kmer_lists[index]\n",
    "    filepath = filepaths[index]\n",
    "    symbol_dict = symbol_dicts[index]\n",
    "    ext = file_exts[index]\n",
    "    print (index,kmers,filepath,symbol_dict)\n",
    "    \n",
    "    if features != 'PC': #PC same sequence as for AA\n",
    "        filepath = filepaths[index]\n",
    "        print ( f'getting {feature} sequences from {filepath}')\n",
    "        sequences = get_sequences(filepath,datasets,ext)\n",
    "    \n",
    "    for k in kmers: \n",
    "        start_time = timeit.default_timer()\n",
    "        x_train,x_test,y_train,y_test = get_feature_matrices(sequences,datasets,k,symbol_dict)\n",
    "        print ('timeit: ',timeit.default_timer() - start_time)\n",
    "        print (np.shape(x_train), np.shape(x_test),np.shape(y_train),np.shape(y_test))\n",
    "        results = fp.test_prediction(x_train,x_test,y_train,y_test)\n",
    "        results.update ({'N in class': n, 'Features':feature, 'k':k})\n",
    "        print(results)\n",
    "        fp.results2CSV (results,subset, results_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
