{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "import os\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import random\n",
    "import numpy as np\n",
    "import timeit\n",
    "from collections import defaultdict, Counter\n",
    "from pathlib import Path\n",
    "from Bio import SeqIO\n",
    "import sys\n",
    "lib_dir = '../mylibs'\n",
    "if lib_dir not in sys.path:\n",
    "    sys.path.append(lib_dir)\n",
    "    \n",
    "import vhdb as vhdb\n",
    "import features_predict as fp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "filein = '../inputs/VHDB_25_1_2019.p'\n",
    "with open(filein, 'rb') as f:\n",
    "    V_H = pickle.load( f)\n",
    "hosts = V_H.hosts\n",
    "viruses = V_H.viruses\n",
    "subspecies =V_H.subspecies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results in Eukaryote_DNA_results.csv\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Input file\n",
    "#subsets = [(label,taxlevel,pool,taxlevelpool),...]\n",
    "subsetfile = '../inputs/Eukaryote_DNA.csv'\n",
    "baltimore = 'all'\n",
    "vhdbfile = '../inputs/VHDB_25_1_2019.p'\n",
    "\n",
    "\n",
    "# Everthing in the same order as the features list\n",
    "\n",
    "feature_list = ['DNA','AA','PC','Domains']\n",
    "# filepath for the fasta, faa and domain files\n",
    "filepaths = [\n",
    "                '/home4/youn01f/Desktop/workspace/newData/fasta',\n",
    "                '/home4/youn01f/Desktop/workspace/newData/faa','/home4/youn01f/Desktop/workspace/newData/faa',\n",
    "                '/home4/youn01f/Desktop/workspace/newData/pfs']\n",
    "file_exts = ['fasta','fasta','fasta','pfs']#,'fasta',fasta]\n",
    "\n",
    "# A list of kmers for each feature set to be tested\n",
    "kmer_lists = [[1, 2,3,4,5,6,7,8,9], # dna \n",
    "              [1, 2, 3,4], # aa\n",
    "              [1, 2,3,4,5,6] ,    #pc\n",
    "             [0]] #domains\n",
    "\n",
    "# symbol dictionaries \n",
    "na_dict = {'mod':4,'a':0,'c':1,'g':2,'t':3}\n",
    "aa_dict = {'mod':20 ,'A':0,'C':1,'D':2,'E':3,'F':4,'G':5,'H':6,'I':7,'K':8,'L':9,'M':10,'N':11,\n",
    "              'P':12,'Q':13,'R':14,'S':15,'T':16,'V':17,'W':18,'Y':19}\n",
    "pc_dict = {'mod':7, 'C':0,\n",
    "            'A':1,'G':1,'V':1,\n",
    "           'I':2,'L':2,'F':2,'P':2,\n",
    "           'M':3,'S':3,'T':3,'Y':3,\n",
    "           'H':4,'N':4,'Q':4,'W':4,\n",
    "           'R':5,'K':5,\n",
    "           'D':6,'E':6}\n",
    "symbol_dicts = [ na_dict,aa_dict,pc_dict, {}]\n",
    "\n",
    "classifiers = ['SVM_lin']\n",
    "\n",
    "# Output file for the results\n",
    "\n",
    "results_file = f'{Path(subsetfile).stem}_results.csv'\n",
    "print( f'Results in {results_file}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "37"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "label_info = pd.read_csv(subsetfile)\n",
    "subsets = label_info.apply(tuple, axis=1).tolist()\n",
    "len(subsets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " size of training 118 and test set   30\n",
      " size of training 120 and test set   32\n",
      " size of training 159 and test set   37\n",
      " size of training 133 and test set   33\n",
      " size of training 133 and test set   33\n",
      " size of training 48 and test set   16\n",
      " size of training 54 and test set   18\n",
      " size of training 185 and test set   47\n",
      " size of training 44 and test set   16\n",
      " size of training 50 and test set   16\n",
      " size of training 187 and test set   47\n",
      " size of training 76 and test set   20\n",
      " size of training 90 and test set   18\n",
      " size of training 402 and test set   98\n",
      " size of training 402 and test set   98\n",
      " size of training 402 and test set   98\n",
      " size of training 44 and test set   16\n",
      " size of training 101 and test set   21\n",
      " size of training 157 and test set   35\n",
      " size of training 402 and test set   98\n",
      " size of training 46 and test set   16\n",
      " size of training 46 and test set   16\n",
      " size of training 96 and test set   20\n",
      " size of training 134 and test set   34\n",
      " size of training 402 and test set   98\n",
      " size of training 52 and test set   16\n",
      " size of training 72 and test set   20\n",
      " size of training 76 and test set   20\n",
      " size of training 196 and test set   50\n",
      " size of training 402 and test set   98\n",
      " size of training 67 and test set   19\n",
      " size of training 189 and test set   47\n",
      " size of training 44 and test set   16\n",
      " size of training 78 and test set   20\n",
      " size of training 370 and test set   88\n",
      " size of training 402 and test set   98\n",
      " size of training 60 and test set   18\n"
     ]
    }
   ],
   "source": [
    "for subset in subsets:\n",
    "    class_data,n = fp.get_class_lists(subset,viruses,hosts)\n",
    "    datasets = fp.split_data(viruses,class_data,n)\n",
    "    (label,label_tax,pool,pool_tax,baltimore) = subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "min(500,250,251)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_kmers (sequences,k,symbol_dict):\n",
    "    if k == 0: # domainsdom_list = list({dom for s in sequence.values() for dom in s})\n",
    "        \n",
    "        X,seq_index,f_index = get_domains(sequences,symbol_dict)\n",
    "    else:  # everything else\n",
    "        vocab_len =  symbol_dict['mod'] **k\n",
    "        f_index = get_feature_names(symbol_dict,k)\n",
    "        seq_index = []\n",
    "        X = np.zeros((len(sequences) ,vocab_len))\n",
    "\n",
    "        for i,(accn, seq) in enumerate (sequences.items()):\n",
    "            X[i] = get_word_frequ (seq,k, vocab_len,symbol_dict)\n",
    "            seq_index.append(accn)\n",
    "# normalise for length of genome(s)\n",
    "    XN = np.divide(X,X.sum(axis = 1)[:,None])\n",
    "    print (np.shape(XN)) \n",
    "    return XN,seq_index,f_index\n",
    "\n",
    "def get_domains(sequence,dom_list):\n",
    "    ''' input: sequences dictionary of training and test sequence dicts'''\n",
    "    seq_index = []\n",
    "    X = np.zeros((len(sequence) ,len(dom_list)))\n",
    "    for i,(accn, seq) in enumerate (sequence.items()):\n",
    "        for dom in seq:\n",
    "            j = dom_list.index(dom)\n",
    "            X[i,j] += 1\n",
    "        seq_index.append(accn)\n",
    "    return X,seq_index,dom_list\n",
    "\n",
    "def get_word_frequ(sequence,k,vocab_len,symbol_dict):\n",
    "    \"\"\"returns the word frequency array for all genomes for a virus \"\"\"\n",
    "    word_freqs = np.zeros(shape =(vocab_len),dtype=np.int32)\n",
    "    for i in range(len(sequence)-k+1):\n",
    "        word = sequence[i:i+k]\n",
    "        if check_word(word,symbol_dict):\n",
    "            kmer_index = patternToNumber (word,k, symbol_dict)\n",
    "            word_freqs [ kmer_index] += 1\n",
    "            # If this for DNA/RNA then add the k-mers complinent\n",
    "#             if symbol_dict['mod'] == 4:\n",
    "#                 word_freqs [patternToNumber(compliment(word),k,symbol_dict)] +=1   \n",
    "    return word_freqs\n",
    "\n",
    "def get_sequences(faapath,datasets,ext):\n",
    "    missing_files = []\n",
    "    ext = f'.{ext}'\n",
    "    sequences = {'training':{},'test':{}}\n",
    "    for key,dataset in datasets.items():\n",
    "        for v,vdict in dataset.items():\n",
    "            sequence = defaultdict(str)\n",
    "            for ref in vdict['refseqs']:\n",
    "                filename = os.path.join(faapath, ref.strip()+ ext)\n",
    "                try:\n",
    "                    with open(filename, 'rt') as handle:\n",
    "                        if ext == '.fasta':\n",
    "                            for record in SeqIO.parse(handle, \"fasta\"):                   \n",
    "                                sequence[v] += str(record.seq)\n",
    "                        else: # pfs file - list of domains  \n",
    "                            line = handle.read()\n",
    "                            sequence[v] = [d for d in line.split()]\n",
    "                                \n",
    "                except IOError:\n",
    "                    missing_files.append(ref)\n",
    "            sequences[key].update(sequence)\n",
    "    print ( f'missing files {len(missing_files)}')\n",
    "    return (sequences)\n",
    "\n",
    "\n",
    "def symbolToNumber (symbol,symbols):\n",
    "    return symbols[symbol]\n",
    "\n",
    "def patternToNumber (pattern,k,symbol_dict):\n",
    "    mod = symbol_dict['mod']\n",
    "    number = 0\n",
    "    for i in range (0,k):\n",
    "        n = symbolToNumber ( pattern [k-1-i],symbol_dict)\n",
    "        number += n* mod**i \n",
    "    return number\n",
    "\n",
    "def get_feature_names(symbols,k):\n",
    "    all_kmers = []\n",
    "    n2sym = { v:k for k,v in symbols.items()}\n",
    "    mod = symbols['mod']\n",
    "    for number in range(mod**k):\n",
    "        pattern = ''\n",
    "        for i in range (0,k):\n",
    "            n = number% mod\n",
    "            number = (number - n)/mod \n",
    "            symbol = n2sym[n]\n",
    "            pattern = symbol + pattern\n",
    "        all_kmers.append (pattern)\n",
    "    return all_kmers\n",
    "\n",
    "def check_word(mystring,symbols):\n",
    "    \"\"\"Check for illegal characters \"\"\"\n",
    "    return all(c in symbols for c in mystring)\n",
    "\n",
    "def compliment(dna):\n",
    "    \"\"\"Return the reverse compliment of a sequence \"\"\"\n",
    "    bp = { 'a':'t', 't':'a', 'c':'g','g':'c'}\n",
    "    revdna =''    \n",
    "    for i in range(len(dna)-1,-1,-1):\n",
    "         revdna += bp[dna[i]]\n",
    "    return (revdna)\n",
    "\n",
    "def  get_labels(label,v_index,dataset):      \n",
    "    for i,v in enumerate(v_index):\n",
    "        y = [1 if dataset[v]['label'] ==  label else 0 for v in v_index ] \n",
    "    return np.asarray (y)\n",
    "\n",
    "def get_feature_matrices(sequences,datasets,k,symbol_dict):\n",
    "   \n",
    "    if k == 0: # domainsdom_list = list({dom for s in sequence.values() for dom in s})\n",
    "        symbol_dict = list({dom for sequence in sequences.values() for s in sequence.values() for dom in s})\n",
    "        \n",
    "    \n",
    "   \n",
    "   # print (f'X_training:Extacting  {feature}  of length  {k}  from   {len(sequences[\"training\"])} sequences')\n",
    "   # start_time = timeit.default_timer()\n",
    "    X_train, seq_index, f_index = extract_kmers (sequences['training'],k,symbol_dict)\n",
    "    Y_train =  get_labels (label,seq_index,datasets['training'])\n",
    "   # print ('timeit: ',timeit.default_timer() - start_time)\n",
    "   # print (f'X_test: Extacting  {feature}  of length  {k}  from   {len(sequences[\"test\"])} sequences')\n",
    "   # start_time = timeit.default_timer()\n",
    "    X_test, seq_index, f_index = extract_kmers (sequences['test'],k,symbol_dict)\n",
    "    Y_test =  get_labels (label,seq_index,datasets['test'])\n",
    "    #print ('timit: ', timeit.default_timer() - start_time)\n",
    "\n",
    "    return X_train,X_test,Y_train,Y_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " size of training 118 and test set   30\n",
      "Bos taurus species Mammalia class DNA\n",
      "0 [1, 2, 3, 4, 5, 6, 7, 8, 9] /home4/youn01f/Desktop/workspace/newData/fasta {'mod': 4, 'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
      "getting DNA sequences from /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "missing files 2\n",
      "(116, 4)\n",
      "(30, 4)\n",
      "(116, 4) (30, 4) (116,) (30,)\n",
      "{'AUC': 0.538, 'N in class': 74, 'Features': 'DNA', 'k': 1}\n",
      "(116, 16)\n",
      "(30, 16)\n",
      "(116, 16) (30, 16) (116,) (30,)\n",
      "{'AUC': 0.778, 'N in class': 74, 'Features': 'DNA', 'k': 2}\n",
      "(116, 64)\n",
      "(30, 64)\n",
      "(116, 64) (30, 64) (116,) (30,)\n",
      "{'AUC': 0.827, 'N in class': 74, 'Features': 'DNA', 'k': 3}\n",
      "(116, 256)\n",
      "(30, 256)\n",
      "(116, 256) (30, 256) (116,) (30,)\n",
      "{'AUC': 0.924, 'N in class': 74, 'Features': 'DNA', 'k': 4}\n",
      "(116, 1024)\n",
      "(30, 1024)\n",
      "(116, 1024) (30, 1024) (116,) (30,)\n",
      "{'AUC': 0.907, 'N in class': 74, 'Features': 'DNA', 'k': 5}\n",
      "(116, 4096)\n",
      "(30, 4096)\n",
      "(116, 4096) (30, 4096) (116,) (30,)\n",
      "{'AUC': 0.929, 'N in class': 74, 'Features': 'DNA', 'k': 6}\n",
      "(116, 16384)\n",
      "(30, 16384)\n",
      "(116, 16384) (30, 16384) (116,) (30,)\n",
      "{'AUC': 0.898, 'N in class': 74, 'Features': 'DNA', 'k': 7}\n",
      "(116, 65536)\n",
      "(30, 65536)\n",
      "(116, 65536) (30, 65536) (116,) (30,)\n",
      "{'AUC': 0.871, 'N in class': 74, 'Features': 'DNA', 'k': 8}\n",
      "(116, 262144)\n",
      "(30, 262144)\n",
      "(116, 262144) (30, 262144) (116,) (30,)\n",
      "{'AUC': 0.818, 'N in class': 74, 'Features': 'DNA', 'k': 9}\n",
      "1 [1, 2, 3, 4] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 20, 'A': 0, 'C': 1, 'D': 2, 'E': 3, 'F': 4, 'G': 5, 'H': 6, 'I': 7, 'K': 8, 'L': 9, 'M': 10, 'N': 11, 'P': 12, 'Q': 13, 'R': 14, 'S': 15, 'T': 16, 'V': 17, 'W': 18, 'Y': 19}\n",
      "getting AA sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "missing files 2\n",
      "(118, 20)\n",
      "(29, 20)\n",
      "(118, 20) (29, 20) (118,) (29,)\n",
      "{'AUC': 0.667, 'N in class': 74, 'Features': 'AA', 'k': 1}\n",
      "(118, 400)\n",
      "(29, 400)\n",
      "(118, 400) (29, 400) (118,) (29,)\n",
      "{'AUC': 0.762, 'N in class': 74, 'Features': 'AA', 'k': 2}\n",
      "(118, 8000)\n",
      "(29, 8000)\n",
      "(118, 8000) (29, 8000) (118,) (29,)\n",
      "{'AUC': 0.824, 'N in class': 74, 'Features': 'AA', 'k': 3}\n",
      "(118, 160000)\n",
      "(29, 160000)\n",
      "(118, 160000) (29, 160000) (118,) (29,)\n",
      "{'AUC': 0.886, 'N in class': 74, 'Features': 'AA', 'k': 4}\n",
      "2 [1, 2, 3, 4, 5, 6] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 7, 'C': 0, 'A': 1, 'G': 1, 'V': 1, 'I': 2, 'L': 2, 'F': 2, 'P': 2, 'M': 3, 'S': 3, 'T': 3, 'Y': 3, 'H': 4, 'N': 4, 'Q': 4, 'W': 4, 'R': 5, 'K': 5, 'D': 6, 'E': 6}\n",
      "getting PC sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "missing files 2\n",
      "(118, 7)\n",
      "(29, 7)\n",
      "(118, 7) (29, 7) (118,) (29,)\n",
      "{'AUC': 0.624, 'N in class': 74, 'Features': 'PC', 'k': 1}\n",
      "(118, 49)\n",
      "(29, 49)\n",
      "(118, 49) (29, 49) (118,) (29,)\n",
      "{'AUC': 0.8, 'N in class': 74, 'Features': 'PC', 'k': 2}\n",
      "(118, 343)\n",
      "(29, 343)\n",
      "(118, 343) (29, 343) (118,) (29,)\n",
      "{'AUC': 0.714, 'N in class': 74, 'Features': 'PC', 'k': 3}\n",
      "(118, 2401)\n",
      "(29, 2401)\n",
      "(118, 2401) (29, 2401) (118,) (29,)\n",
      "{'AUC': 0.752, 'N in class': 74, 'Features': 'PC', 'k': 4}\n",
      "(118, 16807)\n",
      "(29, 16807)\n",
      "(118, 16807) (29, 16807) (118,) (29,)\n",
      "{'AUC': 0.829, 'N in class': 74, 'Features': 'PC', 'k': 5}\n",
      "(118, 117649)\n",
      "(29, 117649)\n",
      "(118, 117649) (29, 117649) (118,) (29,)\n",
      "{'AUC': 0.79, 'N in class': 74, 'Features': 'PC', 'k': 6}\n",
      "3 [0] /home4/youn01f/Desktop/workspace/newData/pfs {}\n",
      "getting Domains sequences from /home4/youn01f/Desktop/workspace/newData/pfs\n",
      "missing files 12\n",
      "(113, 857)\n",
      "(25, 857)\n",
      "(113, 857) (25, 857) (113,) (25,)\n",
      "{'AUC': 0.942, 'N in class': 74, 'Features': 'Domains', 'k': 0}\n",
      " size of training 120 and test set   32\n",
      "Bos genus Mammalia class DNA\n",
      "0 [1, 2, 3, 4, 5, 6, 7, 8, 9] /home4/youn01f/Desktop/workspace/newData/fasta {'mod': 4, 'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
      "getting DNA sequences from /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "missing files 2\n",
      "(118, 4)\n",
      "(32, 4)\n",
      "(118, 4) (32, 4) (118,) (32,)\n",
      "{'AUC': 0.531, 'N in class': 76, 'Features': 'DNA', 'k': 1}\n",
      "(118, 16)\n",
      "(32, 16)\n",
      "(118, 16) (32, 16) (118,) (32,)\n",
      "{'AUC': 0.792, 'N in class': 76, 'Features': 'DNA', 'k': 2}\n",
      "(118, 64)\n",
      "(32, 64)\n",
      "(118, 64) (32, 64) (118,) (32,)\n",
      "{'AUC': 0.925, 'N in class': 76, 'Features': 'DNA', 'k': 3}\n",
      "(118, 256)\n",
      "(32, 256)\n",
      "(118, 256) (32, 256) (118,) (32,)\n",
      "{'AUC': 0.953, 'N in class': 76, 'Features': 'DNA', 'k': 4}\n",
      "(118, 1024)\n",
      "(32, 1024)\n",
      "(118, 1024) (32, 1024) (118,) (32,)\n",
      "{'AUC': 0.98, 'N in class': 76, 'Features': 'DNA', 'k': 5}\n",
      "(118, 4096)\n",
      "(32, 4096)\n",
      "(118, 4096) (32, 4096) (118,) (32,)\n",
      "{'AUC': 0.982, 'N in class': 76, 'Features': 'DNA', 'k': 6}\n",
      "(118, 16384)\n",
      "(32, 16384)\n",
      "(118, 16384) (32, 16384) (118,) (32,)\n",
      "{'AUC': 0.976, 'N in class': 76, 'Features': 'DNA', 'k': 7}\n",
      "(118, 65536)\n",
      "(32, 65536)\n",
      "(118, 65536) (32, 65536) (118,) (32,)\n",
      "{'AUC': 0.945, 'N in class': 76, 'Features': 'DNA', 'k': 8}\n",
      "(118, 262144)\n",
      "(32, 262144)\n",
      "(118, 262144) (32, 262144) (118,) (32,)\n",
      "{'AUC': 0.941, 'N in class': 76, 'Features': 'DNA', 'k': 9}\n",
      "1 [1, 2, 3, 4] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 20, 'A': 0, 'C': 1, 'D': 2, 'E': 3, 'F': 4, 'G': 5, 'H': 6, 'I': 7, 'K': 8, 'L': 9, 'M': 10, 'N': 11, 'P': 12, 'Q': 13, 'R': 14, 'S': 15, 'T': 16, 'V': 17, 'W': 18, 'Y': 19}\n",
      "getting AA sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "missing files 2\n",
      "(118, 20)\n",
      "(32, 20)\n",
      "(118, 20) (32, 20) (118,) (32,)\n",
      "{'AUC': 0.967, 'N in class': 76, 'Features': 'AA', 'k': 1}\n",
      "(118, 400)\n",
      "(32, 400)\n",
      "(118, 400) (32, 400) (118,) (32,)\n",
      "{'AUC': 0.957, 'N in class': 76, 'Features': 'AA', 'k': 2}\n",
      "(118, 8000)\n",
      "(32, 8000)\n",
      "(118, 8000) (32, 8000) (118,) (32,)\n",
      "{'AUC': 0.91, 'N in class': 76, 'Features': 'AA', 'k': 3}\n",
      "(118, 160000)\n",
      "(32, 160000)\n",
      "(118, 160000) (32, 160000) (118,) (32,)\n",
      "{'AUC': 0.937, 'N in class': 76, 'Features': 'AA', 'k': 4}\n",
      "2 [1, 2, 3, 4, 5, 6] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 7, 'C': 0, 'A': 1, 'G': 1, 'V': 1, 'I': 2, 'L': 2, 'F': 2, 'P': 2, 'M': 3, 'S': 3, 'T': 3, 'Y': 3, 'H': 4, 'N': 4, 'Q': 4, 'W': 4, 'R': 5, 'K': 5, 'D': 6, 'E': 6}\n",
      "getting PC sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "missing files 2\n",
      "(118, 7)\n",
      "(32, 7)\n",
      "(118, 7) (32, 7) (118,) (32,)\n",
      "{'AUC': 0.898, 'N in class': 76, 'Features': 'PC', 'k': 1}\n",
      "(118, 49)\n",
      "(32, 49)\n",
      "(118, 49) (32, 49) (118,) (32,)\n",
      "{'AUC': 0.855, 'N in class': 76, 'Features': 'PC', 'k': 2}\n",
      "(118, 343)\n",
      "(32, 343)\n",
      "(118, 343) (32, 343) (118,) (32,)\n",
      "{'AUC': 0.949, 'N in class': 76, 'Features': 'PC', 'k': 3}\n",
      "(118, 2401)\n",
      "(32, 2401)\n",
      "(118, 2401) (32, 2401) (118,) (32,)\n",
      "{'AUC': 0.906, 'N in class': 76, 'Features': 'PC', 'k': 4}\n",
      "(118, 16807)\n",
      "(32, 16807)\n",
      "(118, 16807) (32, 16807) (118,) (32,)\n",
      "{'AUC': 0.953, 'N in class': 76, 'Features': 'PC', 'k': 5}\n",
      "(118, 117649)\n",
      "(32, 117649)\n",
      "(118, 117649) (32, 117649) (118,) (32,)\n",
      "{'AUC': 0.918, 'N in class': 76, 'Features': 'PC', 'k': 6}\n",
      "3 [0] /home4/youn01f/Desktop/workspace/newData/pfs {}\n",
      "getting Domains sequences from /home4/youn01f/Desktop/workspace/newData/pfs\n",
      "missing files 11\n",
      "(110, 1038)\n",
      "(31, 1038)\n",
      "(110, 1038) (31, 1038) (110,) (31,)\n",
      "{'AUC': 0.988, 'N in class': 76, 'Features': 'Domains', 'k': 0}\n",
      " size of training 159 and test set   37\n",
      "Bovidae family Mammalia class DNA\n",
      "0 [1, 2, 3, 4, 5, 6, 7, 8, 9] /home4/youn01f/Desktop/workspace/newData/fasta {'mod': 4, 'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
      "getting DNA sequences from /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "missing files 3\n",
      "(157, 4)\n",
      "(36, 4)\n",
      "(157, 4) (36, 4) (157,) (36,)\n",
      "{'AUC': 0.412, 'N in class': 98, 'Features': 'DNA', 'k': 1}\n",
      "(157, 16)\n",
      "(36, 16)\n",
      "(157, 16) (36, 16) (157,) (36,)\n",
      "{'AUC': 0.669, 'N in class': 98, 'Features': 'DNA', 'k': 2}\n",
      "(157, 64)\n",
      "(36, 64)\n",
      "(157, 64) (36, 64) (157,) (36,)\n",
      "{'AUC': 0.813, 'N in class': 98, 'Features': 'DNA', 'k': 3}\n",
      "(157, 256)\n",
      "(36, 256)\n",
      "(157, 256) (36, 256) (157,) (36,)\n",
      "{'AUC': 0.841, 'N in class': 98, 'Features': 'DNA', 'k': 4}\n",
      "(157, 1024)\n",
      "(36, 1024)\n",
      "(157, 1024) (36, 1024) (157,) (36,)\n",
      "{'AUC': 0.884, 'N in class': 98, 'Features': 'DNA', 'k': 5}\n",
      "(157, 4096)\n",
      "(36, 4096)\n",
      "(157, 4096) (36, 4096) (157,) (36,)\n",
      "{'AUC': 0.869, 'N in class': 98, 'Features': 'DNA', 'k': 6}\n",
      "(157, 16384)\n",
      "(36, 16384)\n",
      "(157, 16384) (36, 16384) (157,) (36,)\n",
      "{'AUC': 0.853, 'N in class': 98, 'Features': 'DNA', 'k': 7}\n",
      "(157, 65536)\n",
      "(36, 65536)\n",
      "(157, 65536) (36, 65536) (157,) (36,)\n",
      "{'AUC': 0.837, 'N in class': 98, 'Features': 'DNA', 'k': 8}\n",
      "(157, 262144)\n",
      "(36, 262144)\n",
      "(157, 262144) (36, 262144) (157,) (36,)\n",
      "{'AUC': 0.819, 'N in class': 98, 'Features': 'DNA', 'k': 9}\n",
      "1 [1, 2, 3, 4] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 20, 'A': 0, 'C': 1, 'D': 2, 'E': 3, 'F': 4, 'G': 5, 'H': 6, 'I': 7, 'K': 8, 'L': 9, 'M': 10, 'N': 11, 'P': 12, 'Q': 13, 'R': 14, 'S': 15, 'T': 16, 'V': 17, 'W': 18, 'Y': 19}\n",
      "getting AA sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "missing files 3\n",
      "(156, 20)\n",
      "(37, 20)\n",
      "(156, 20) (37, 20) (156,) (37,)\n",
      "{'AUC': 0.718, 'N in class': 98, 'Features': 'AA', 'k': 1}\n",
      "(156, 400)\n",
      "(37, 400)\n",
      "(156, 400) (37, 400) (156,) (37,)\n",
      "{'AUC': 0.85, 'N in class': 98, 'Features': 'AA', 'k': 2}\n",
      "(156, 8000)\n",
      "(37, 8000)\n",
      "(156, 8000) (37, 8000) (156,) (37,)\n",
      "{'AUC': 0.794, 'N in class': 98, 'Features': 'AA', 'k': 3}\n",
      "(156, 160000)\n",
      "(37, 160000)\n",
      "(156, 160000) (37, 160000) (156,) (37,)\n",
      "{'AUC': 0.753, 'N in class': 98, 'Features': 'AA', 'k': 4}\n",
      "2 [1, 2, 3, 4, 5, 6] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 7, 'C': 0, 'A': 1, 'G': 1, 'V': 1, 'I': 2, 'L': 2, 'F': 2, 'P': 2, 'M': 3, 'S': 3, 'T': 3, 'Y': 3, 'H': 4, 'N': 4, 'Q': 4, 'W': 4, 'R': 5, 'K': 5, 'D': 6, 'E': 6}\n",
      "getting PC sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "missing files 3\n",
      "(156, 7)\n",
      "(37, 7)\n",
      "(156, 7) (37, 7) (156,) (37,)\n",
      "{'AUC': 0.603, 'N in class': 98, 'Features': 'PC', 'k': 1}\n",
      "(156, 49)\n",
      "(37, 49)\n",
      "(156, 49) (37, 49) (156,) (37,)\n",
      "{'AUC': 0.691, 'N in class': 98, 'Features': 'PC', 'k': 2}\n",
      "(156, 343)\n",
      "(37, 343)\n",
      "(156, 343) (37, 343) (156,) (37,)\n",
      "{'AUC': 0.829, 'N in class': 98, 'Features': 'PC', 'k': 3}\n",
      "(156, 2401)\n",
      "(37, 2401)\n",
      "(156, 2401) (37, 2401) (156,) (37,)\n",
      "{'AUC': 0.856, 'N in class': 98, 'Features': 'PC', 'k': 4}\n",
      "(156, 16807)\n",
      "(37, 16807)\n",
      "(156, 16807) (37, 16807) (156,) (37,)\n",
      "{'AUC': 0.791, 'N in class': 98, 'Features': 'PC', 'k': 5}\n",
      "(156, 117649)\n",
      "(37, 117649)\n",
      "(156, 117649) (37, 117649) (156,) (37,)\n",
      "{'AUC': 0.765, 'N in class': 98, 'Features': 'PC', 'k': 6}\n",
      "3 [0] /home4/youn01f/Desktop/workspace/newData/pfs {}\n",
      "getting Domains sequences from /home4/youn01f/Desktop/workspace/newData/pfs\n",
      "missing files 15\n",
      "(146, 1078)\n",
      "(35, 1078)\n",
      "(146, 1078) (35, 1078) (146,) (35,)\n",
      "{'AUC': 0.918, 'N in class': 98, 'Features': 'Domains', 'k': 0}\n",
      " size of training 133 and test set   33\n",
      "Sus scrofa species Mammalia class DNA\n",
      "0 [1, 2, 3, 4, 5, 6, 7, 8, 9] /home4/youn01f/Desktop/workspace/newData/fasta {'mod': 4, 'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
      "getting DNA sequences from /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "missing files 0\n",
      "(133, 4)\n",
      "(33, 4)\n",
      "(133, 4) (33, 4) (133,) (33,)\n",
      "{'AUC': 0.43, 'N in class': 83, 'Features': 'DNA', 'k': 1}\n",
      "(133, 16)\n",
      "(33, 16)\n",
      "(133, 16) (33, 16) (133,) (33,)\n",
      "{'AUC': 0.864, 'N in class': 83, 'Features': 'DNA', 'k': 2}\n",
      "(133, 64)\n",
      "(33, 64)\n",
      "(133, 64) (33, 64) (133,) (33,)\n",
      "{'AUC': 0.824, 'N in class': 83, 'Features': 'DNA', 'k': 3}\n",
      "(133, 256)\n",
      "(33, 256)\n",
      "(133, 256) (33, 256) (133,) (33,)\n",
      "{'AUC': 0.93, 'N in class': 83, 'Features': 'DNA', 'k': 4}\n",
      "(133, 1024)\n",
      "(33, 1024)\n",
      "(133, 1024) (33, 1024) (133,) (33,)\n",
      "{'AUC': 0.941, 'N in class': 83, 'Features': 'DNA', 'k': 5}\n",
      "(133, 4096)\n",
      "(33, 4096)\n",
      "(133, 4096) (33, 4096) (133,) (33,)\n",
      "{'AUC': 0.934, 'N in class': 83, 'Features': 'DNA', 'k': 6}\n",
      "(133, 16384)\n",
      "(33, 16384)\n",
      "(133, 16384) (33, 16384) (133,) (33,)\n",
      "{'AUC': 0.934, 'N in class': 83, 'Features': 'DNA', 'k': 7}\n",
      "(133, 65536)\n",
      "(33, 65536)\n",
      "(133, 65536) (33, 65536) (133,) (33,)\n",
      "{'AUC': 0.945, 'N in class': 83, 'Features': 'DNA', 'k': 8}\n",
      "(133, 262144)\n",
      "(33, 262144)\n",
      "(133, 262144) (33, 262144) (133,) (33,)\n",
      "{'AUC': 0.919, 'N in class': 83, 'Features': 'DNA', 'k': 9}\n",
      "1 [1, 2, 3, 4] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 20, 'A': 0, 'C': 1, 'D': 2, 'E': 3, 'F': 4, 'G': 5, 'H': 6, 'I': 7, 'K': 8, 'L': 9, 'M': 10, 'N': 11, 'P': 12, 'Q': 13, 'R': 14, 'S': 15, 'T': 16, 'V': 17, 'W': 18, 'Y': 19}\n",
      "getting AA sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "missing files 3\n",
      "(130, 20)\n",
      "(33, 20)\n",
      "(130, 20) (33, 20) (130,) (33,)\n",
      "{'AUC': 0.846, 'N in class': 83, 'Features': 'AA', 'k': 1}\n",
      "(130, 400)\n",
      "(33, 400)\n",
      "(130, 400) (33, 400) (130,) (33,)\n",
      "{'AUC': 0.938, 'N in class': 83, 'Features': 'AA', 'k': 2}\n",
      "(130, 8000)\n",
      "(33, 8000)\n",
      "(130, 8000) (33, 8000) (130,) (33,)\n",
      "{'AUC': 0.941, 'N in class': 83, 'Features': 'AA', 'k': 3}\n",
      "(130, 160000)\n",
      "(33, 160000)\n",
      "(130, 160000) (33, 160000) (130,) (33,)\n",
      "{'AUC': 0.952, 'N in class': 83, 'Features': 'AA', 'k': 4}\n",
      "2 [1, 2, 3, 4, 5, 6] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 7, 'C': 0, 'A': 1, 'G': 1, 'V': 1, 'I': 2, 'L': 2, 'F': 2, 'P': 2, 'M': 3, 'S': 3, 'T': 3, 'Y': 3, 'H': 4, 'N': 4, 'Q': 4, 'W': 4, 'R': 5, 'K': 5, 'D': 6, 'E': 6}\n",
      "getting PC sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "missing files 3\n",
      "(130, 7)\n",
      "(33, 7)\n",
      "(130, 7) (33, 7) (130,) (33,)\n",
      "{'AUC': 0.739, 'N in class': 83, 'Features': 'PC', 'k': 1}\n",
      "(130, 49)\n",
      "(33, 49)\n",
      "(130, 49) (33, 49) (130,) (33,)\n",
      "{'AUC': 0.779, 'N in class': 83, 'Features': 'PC', 'k': 2}\n",
      "(130, 343)\n",
      "(33, 343)\n",
      "(130, 343) (33, 343) (130,) (33,)\n",
      "{'AUC': 0.838, 'N in class': 83, 'Features': 'PC', 'k': 3}\n",
      "(130, 2401)\n",
      "(33, 2401)\n",
      "(130, 2401) (33, 2401) (130,) (33,)\n",
      "{'AUC': 0.93, 'N in class': 83, 'Features': 'PC', 'k': 4}\n",
      "(130, 16807)\n",
      "(33, 16807)\n",
      "(130, 16807) (33, 16807) (130,) (33,)\n",
      "{'AUC': 0.96, 'N in class': 83, 'Features': 'PC', 'k': 5}\n",
      "(130, 117649)\n",
      "(33, 117649)\n",
      "(130, 117649) (33, 117649) (130,) (33,)\n",
      "{'AUC': 0.971, 'N in class': 83, 'Features': 'PC', 'k': 6}\n",
      "3 [0] /home4/youn01f/Desktop/workspace/newData/pfs {}\n",
      "getting Domains sequences from /home4/youn01f/Desktop/workspace/newData/pfs\n",
      "missing files 21\n",
      "(115, 1128)\n",
      "(30, 1128)\n",
      "(115, 1128) (30, 1128) (115,) (30,)\n",
      "{'AUC': 0.83, 'N in class': 83, 'Features': 'Domains', 'k': 0}\n",
      " size of training 133 and test set   33\n",
      "Suidae family Mammalia class DNA\n",
      "0 [1, 2, 3, 4, 5, 6, 7, 8, 9] /home4/youn01f/Desktop/workspace/newData/fasta {'mod': 4, 'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
      "getting DNA sequences from /home4/youn01f/Desktop/workspace/newData/fasta\n",
      "missing files 1\n",
      "(132, 4)\n",
      "(33, 4)\n",
      "(132, 4) (33, 4) (132,) (33,)\n",
      "{'AUC': 0.5, 'N in class': 83, 'Features': 'DNA', 'k': 1}\n",
      "(132, 16)\n",
      "(33, 16)\n",
      "(132, 16) (33, 16) (132,) (33,)\n",
      "{'AUC': 0.824, 'N in class': 83, 'Features': 'DNA', 'k': 2}\n",
      "(132, 64)\n",
      "(33, 64)\n",
      "(132, 64) (33, 64) (132,) (33,)\n",
      "{'AUC': 0.643, 'N in class': 83, 'Features': 'DNA', 'k': 3}\n",
      "(132, 256)\n",
      "(33, 256)\n",
      "(132, 256) (33, 256) (132,) (33,)\n",
      "{'AUC': 0.695, 'N in class': 83, 'Features': 'DNA', 'k': 4}\n",
      "(132, 1024)\n",
      "(33, 1024)\n",
      "(132, 1024) (33, 1024) (132,) (33,)\n",
      "{'AUC': 0.688, 'N in class': 83, 'Features': 'DNA', 'k': 5}\n",
      "(132, 4096)\n",
      "(33, 4096)\n",
      "(132, 4096) (33, 4096) (132,) (33,)\n",
      "{'AUC': 0.809, 'N in class': 83, 'Features': 'DNA', 'k': 6}\n",
      "(132, 16384)\n",
      "(33, 16384)\n",
      "(132, 16384) (33, 16384) (132,) (33,)\n",
      "{'AUC': 0.875, 'N in class': 83, 'Features': 'DNA', 'k': 7}\n",
      "(132, 65536)\n",
      "(33, 65536)\n",
      "(132, 65536) (33, 65536) (132,) (33,)\n",
      "{'AUC': 0.923, 'N in class': 83, 'Features': 'DNA', 'k': 8}\n",
      "(132, 262144)\n",
      "(33, 262144)\n",
      "(132, 262144) (33, 262144) (132,) (33,)\n",
      "{'AUC': 0.757, 'N in class': 83, 'Features': 'DNA', 'k': 9}\n",
      "1 [1, 2, 3, 4] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 20, 'A': 0, 'C': 1, 'D': 2, 'E': 3, 'F': 4, 'G': 5, 'H': 6, 'I': 7, 'K': 8, 'L': 9, 'M': 10, 'N': 11, 'P': 12, 'Q': 13, 'R': 14, 'S': 15, 'T': 16, 'V': 17, 'W': 18, 'Y': 19}\n",
      "getting AA sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "missing files 1\n",
      "(133, 20)\n",
      "(32, 20)\n",
      "(133, 20) (32, 20) (133,) (32,)\n",
      "{'AUC': 0.689, 'N in class': 83, 'Features': 'AA', 'k': 1}\n",
      "(133, 400)\n",
      "(32, 400)\n",
      "(133, 400) (32, 400) (133,) (32,)\n",
      "{'AUC': 0.773, 'N in class': 83, 'Features': 'AA', 'k': 2}\n",
      "(133, 8000)\n",
      "(32, 8000)\n",
      "(133, 8000) (32, 8000) (133,) (32,)\n",
      "{'AUC': 0.66, 'N in class': 83, 'Features': 'AA', 'k': 3}\n",
      "(133, 160000)\n",
      "(32, 160000)\n",
      "(133, 160000) (32, 160000) (133,) (32,)\n",
      "{'AUC': 0.82, 'N in class': 83, 'Features': 'AA', 'k': 4}\n",
      "2 [1, 2, 3, 4, 5, 6] /home4/youn01f/Desktop/workspace/newData/faa {'mod': 7, 'C': 0, 'A': 1, 'G': 1, 'V': 1, 'I': 2, 'L': 2, 'F': 2, 'P': 2, 'M': 3, 'S': 3, 'T': 3, 'Y': 3, 'H': 4, 'N': 4, 'Q': 4, 'W': 4, 'R': 5, 'K': 5, 'D': 6, 'E': 6}\n",
      "getting PC sequences from /home4/youn01f/Desktop/workspace/newData/faa\n",
      "missing files 1\n",
      "(133, 7)\n",
      "(32, 7)\n",
      "(133, 7) (32, 7) (133,) (32,)\n",
      "{'AUC': 0.574, 'N in class': 83, 'Features': 'PC', 'k': 1}\n",
      "(133, 49)\n",
      "(32, 49)\n",
      "(133, 49) (32, 49) (133,) (32,)\n",
      "{'AUC': 0.742, 'N in class': 83, 'Features': 'PC', 'k': 2}\n",
      "(133, 343)\n",
      "(32, 343)\n",
      "(133, 343) (32, 343) (133,) (32,)\n",
      "{'AUC': 0.715, 'N in class': 83, 'Features': 'PC', 'k': 3}\n",
      "(133, 2401)\n",
      "(32, 2401)\n",
      "(133, 2401) (32, 2401) (133,) (32,)\n",
      "{'AUC': 0.73, 'N in class': 83, 'Features': 'PC', 'k': 4}\n"
     ]
    }
   ],
   "source": [
    "features = ['DNA','AA','PC','Domains']\n",
    "for subset in subsets:\n",
    "    class_data,n = fp.get_class_lists(subset,viruses,hosts)\n",
    "    datasets = fp.split_data(viruses,class_data,n)\n",
    "    (label,label_tax,pool,pool_tax,baltimore) = subset\n",
    "    print  (label,label_tax,pool,pool_tax,baltimore)\n",
    "   \n",
    "    for feature in features:\n",
    "        \n",
    "        index = feature_list.index(feature)\n",
    "        kmers = kmer_lists[index]\n",
    "        filepath = filepaths[index]\n",
    "        symbol_dict = symbol_dicts[index]\n",
    "        ext = file_exts[index]\n",
    "        print (index,kmers,filepath,symbol_dict)\n",
    "\n",
    "        if features != 'PC': #PC same sequence as for AA\n",
    "            filepath = filepaths[index]\n",
    "            print ( f'getting {feature} sequences from {filepath}')\n",
    "            sequences = get_sequences(filepath,datasets,ext)\n",
    "\n",
    "        for k in kmers: \n",
    "            x_train,x_test,y_train,y_test = get_feature_matrices(sequences,datasets,k,symbol_dict)\n",
    "            print (np.shape(x_train), np.shape(x_test),np.shape(y_train),np.shape(y_test))\n",
    "            #print (type(x_train), type(x_test),type(y_train),type(y_test))\n",
    "            results = fp.test_prediction(x_train,x_test,y_train,y_test)\n",
    "            results.update ({'N in class': n, 'Features':feature, 'k':k})\n",
    "            print(results)\n",
    "            fp.results2CSV (results,subset, results_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_list = ['DNA','AA','PC','Domains']\n",
    "features = ['Domains']\n",
    "for feature in features:\n",
    "    index = feature_list.index(feature)\n",
    "    kmers = kmer_lists[index]\n",
    "    filepath = filepaths[index]\n",
    "    symbol_dict = symbol_dicts[index]\n",
    "    ext = file_exts[index]\n",
    "    print (index,kmers,filepath,symbol_dict)\n",
    "    \n",
    "    if features != 'PC': #PC same sequence as for AA\n",
    "        filepath = filepaths[index]\n",
    "        print ( f'getting {feature} sequences from {filepath}')\n",
    "        sequences = get_sequences(filepath,datasets,ext)\n",
    "    \n",
    "    for k in kmers: \n",
    "        x_train,x_test,y_train,y_test = get_feature_matrices(sequences,datasets,k,symbol_dict)\n",
    "        print (np.shape(x_train), np.shape(x_test),np.shape(y_train),np.shape(y_test))\n",
    "        print (type(x_train), type(x_test),type(y_train),type(y_test))\n",
    "        results = fp.test_prediction(x_train,x_test,y_train,y_test)\n",
    "        results.update ({'N in class': n, 'Features':feature, 'k':k})\n",
    "        print(results)\n",
    "        fp.results2CSV (results,subset, results_file)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
